

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Examples &mdash; gffutils 0.8.7.1 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="gffutils 0.8.7.1 documentation" href="index.html"/>
        <link rel="next" title="API" href="api.html"/>
        <link rel="prev" title="Attributes" href="attributes.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="contents.html" class="icon icon-home"> gffutils
          

          
          </a>

          
            
            
              <div class="version">
                0.8.7.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="index.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="database-import.html">Importing data into a database</a></li>
<li class="toctree-l1"><a class="reference internal" href="database-schema.html">Database schema</a></li>
<li class="toctree-l1"><a class="reference internal" href="database-ids.html">Database IDs</a></li>
<li class="toctree-l1"><a class="reference internal" href="gtf.html">GTF files</a></li>
<li class="toctree-l1"><a class="reference internal" href="dialect.html">Dialects</a></li>
<li class="toctree-l1"><a class="reference internal" href="attributes.html">Attributes</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Examples</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#mouse-extra-comma-gff3">mouse_extra_comma.gff3</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#file-contents">File contents</a></li>
<li class="toctree-l3"><a class="reference internal" href="#import">Import</a></li>
<li class="toctree-l3"><a class="reference internal" href="#example-access">Example access</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#c-elegans-ws199-ann-gff-txt">c_elegans_WS199_ann_gff.txt</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id4">File contents</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id5">Import</a></li>
<li class="toctree-l3"><a class="reference internal" href="#access">Access</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#c-elegans-ws199-shortened-gff-txt">c_elegans_WS199_shortened_gff.txt</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id7">File contents</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id8">Import</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id9">Access</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#ensembl-gtf-txt">ensembl_gtf.txt</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id11">File contents</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id12">Import</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id13">Access</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#f3-unique-3-v2-gff">F3-unique-3.v2.gff</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id15">File contents</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id16">Import</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#glimmer-nokeyval-gff3">glimmer_nokeyval.gff3</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id18">File contents</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id19">Import</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id20">Access</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#hybrid1-gff3">hybrid1.gff3</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id22">File contents</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id23">Import</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id24">Access</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#jgi-gff2-txt">jgi_gff2.txt</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id26">File contents</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id27">Import</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id28">Access</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#ncbi-gff3-txt">ncbi_gff3.txt</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id30">File contents</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id31">Import</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id32">Access</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#wormbase-gff2-alt-txt">wormbase_gff2_alt.txt</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id34">File contents</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id35">Import</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id36">Access</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#wormbase-gff2-txt">wormbase_gff2.txt</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id38">File contents</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id39">Import</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id40">Access</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#gencode-v19-gtf">gencode-v19.gtf</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id41">Import</a></li>
<li class="toctree-l3"><a class="reference internal" href="#id42">Access</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Change log</a></li>
<li class="toctree-l1"><a class="reference internal" href="developers.html">Developer&#8217;s docs</a></li>
<li class="toctree-l1"><a class="reference internal" href="meta.html">Meta-docs (docs about the docs)</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="contents.html">gffutils</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          





<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="contents.html">Docs</a> &raquo;</li>
      
    <li>Examples</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/examples.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="examples">
<span id="id1"></span><h1>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">Â¶</a></h1>
<p>Often, GTF/GFF files have non-standard formatting.  While <a class="reference internal" href="api.html#module-gffutils" title="gffutils"><code class="xref py py-mod docutils literal"><span class="pre">gffutils</span></code></a> is
set up to handle standard formatting by default, it also allows great
flexibility for working with non-standard files.  These examples collect
various real-world, difficult-to-work-with files and show how to make them work
with <a class="reference internal" href="api.html#module-gffutils" title="gffutils"><code class="xref py py-mod docutils literal"><span class="pre">gffutils</span></code></a>.  If you have a troublesome file, the best thing to do is
to look at the first few lines for these examples and see if you can find one
with matching formatting.</p>
<p>These example files ship with <a class="reference internal" href="api.html#module-gffutils" title="gffutils"><code class="xref py py-mod docutils literal"><span class="pre">gffutils</span></code></a>.  For interactive use, their
absolute path can be obtained with the <code class="xref py py-func docutils literal"><span class="pre">gffutils.example_filename()</span></code>
function. Each example shows the contents of the file, along with text
explaining the troublesome parts and which settings to use in order to
successfully import into a database and retrieve features.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>Be mindful of the difference between the ID attribute of a feature and the
primary key of that feature in the database.</p>
<p class="last"><code class="xref py py-attr docutils literal"><span class="pre">Feature.id</span></code> will always return the primary key.  <code class="docutils literal"><span class="pre">Feature[&quot;ID&quot;]</span></code> will
return the ID attribute of the feature, if it has one.  In many (but not
all) cases, <code class="docutils literal"><span class="pre">Feature.id</span> <span class="pre">==</span> <span class="pre">Feature[&quot;ID&quot;]</span></code></p>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Also be mindful of the difference between an attribute of a Python object
vs the 9th field of a GFF line, which is called the &#8220;attributes&#8221; field
according to the spec. Confusing?  Yes!</p>
</div>
<p>This documentation undergoes automated testing, and in order to support both
Python 2 and 3 in the same test suite, we need to do this:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">__future__</span> <span class="kn">import</span> <span class="n">print_function</span>
</pre></div>
</div>
<div class="html-toggle section" id="mouse-extra-comma-gff3">
<span id="id2"></span><h2>mouse_extra_comma.gff3<a class="headerlink" href="#mouse-extra-comma-gff3" title="Permalink to this headline">Â¶</a></h2>
<p>This file has gene, mRNA, protein, UTR, CDS, and exon annotations. However, not
all of them have a useful ID in their attributes.</p>
<div class="section" id="file-contents">
<h3>File contents<a class="headerlink" href="#file-contents" title="Permalink to this headline">Â¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>chr17	RefSeq	gene	6797760	6818159	.	+	.	ID=NC_000083.5:LOC100040603;Name=NC_000083.5:LOC100040603
chr17	RefSeq	mRNA	6797760	6818159	.	+	.	ID=XM_001475631.1;Parent=NC_000083.5:LOC100040603
chr17	RefSeq	protein	6806527	6812289	.	+	.	ID=;Parent=XM_001475631.1
chr17	RefSeq	five_prime_UTR	6797760	6797769	.	+	.	Parent=XM_001475631.1
chr17	RefSeq	five_prime_UTR	6806513	6806526	.	+	.	Parent=XM_001475631.1
chr17	RefSeq	CDS	6806527	6806553	.	+	0	Name=CDS:NC_000083.5:LOC100040603;Parent=XM_001475631.1,
chr17	RefSeq	CDS	6808204	6808245	.	+	0	Name=CDS:NC_000083.5:LOC100040603;Parent=XM_001475631.1,
chr17	RefSeq	CDS	6811330	6811453	.	+	0	Name=CDS:NC_000083.5:LOC100040603;Parent=XM_001475631.1,
chr17	RefSeq	CDS	6811792	6811869	.	+	2	Name=CDS:NC_000083.5:LOC100040603;Parent=XM_001475631.1,
chr17	RefSeq	CDS	6812219	6812289	.	+	2	Name=CDS:NC_000083.5:LOC100040603;Parent=XM_001475631.1,
chr17	RefSeq	three_prime_UTR	6812290	6818159	.	+	.	Parent=XM_001475631.1
chr17	RefSeq	exon	6797760	6797769	.	+	.	Parent=XM_001475631.1
chr17	RefSeq	exon	6806513	6806553	.	+	.	Parent=XM_001475631.1
chr17	RefSeq	exon	6808204	6808245	.	+	.	Parent=XM_001475631.1
chr17	RefSeq	exon	6811330	6811453	.	+	.	Parent=XM_001475631.1
chr17	RefSeq	exon	6811792	6811869	.	+	.	Parent=XM_001475631.1
chr17	RefSeq	exon	6812219	6818159	.	+	.	Parent=XM_001475631.1
</pre></div>
</div>
</div>
<div class="section" id="import">
<h3>Import<a class="headerlink" href="#import" title="Permalink to this headline">Â¶</a></h3>
<p>The defaults for <a class="reference internal" href="api.html#module-gffutils" title="gffutils"><code class="xref py py-mod docutils literal"><span class="pre">gffutils</span></code></a> will load this example into a database.
However, the CDSs do not have ID attributes, and the default for GFF files is
<code class="docutils literal"><span class="pre">id_spec=&quot;ID&quot;</span></code>.  As described at <a class="reference internal" href="database-ids.html#id-spec"><span>id_spec</span></a>, they will be accessed by the
names <code class="docutils literal"><span class="pre">&quot;CDS_1&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;CDS_2&quot;</span></code>, and so on.</p>
<p>Note that using <code class="docutils literal"><span class="pre">id_spec=[&quot;ID&quot;,</span> <span class="pre">&quot;Name&quot;]</span></code> won&#8217;t work, because the CDSs share the
same Name attribute, <code class="docutils literal"><span class="pre">CDS:NC_000083.5:LOC100040603</span></code>. We could use
<code class="docutils literal"><span class="pre">id_spec=[&quot;ID&quot;,</span> <span class="pre">&quot;Name&quot;]</span></code> in combination with <code class="docutils literal"><span class="pre">merge_strategy=&quot;create_unique&quot;</span></code>,
which would give ids like <code class="docutils literal"><span class="pre">CDS:NC_000083.5:LOC100040603_1</span></code> (with an underscore
an integer at the end for each consecutive occurence of the same ID).  But
we&#8217;ll go with the simpler default strategy for this file:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">gffutils</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fn</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span><span class="s">&#39;mouse_extra_comma.gff3&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">create_db</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="s">&#39;:memory:&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="example-access">
<h3>Example access<a class="headerlink" href="#example-access" title="Permalink to this headline">Â¶</a></h3>
<p>Since the mRNA on line 2 has an ID, we can simply access it by ID:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">db</span><span class="p">[</span><span class="s">&#39;XM_001475631.1&#39;</span><span class="p">]</span>
<span class="go">&lt;Feature mRNA (chr17:6797760-6818159[+]) at 0x...&gt;</span>
</pre></div>
</div>
<p>It&#8217;s not very convenient to access the CDSs by their newly created ids
<code class="docutils literal"><span class="pre">&quot;CDS_1&quot;</span></code>, <code class="docutils literal"><span class="pre">&quot;CDS_2&quot;</span></code>, etc.  (Which one was &#8220;1&#8221;?  Is that from the same gene as
CDS &#8220;5&#8221;?) But we can access those CDSs as children of the mRNA:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">f</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">db</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="s">&#39;XM_001475631.1&#39;</span><span class="p">,</span> <span class="n">featuretype</span><span class="o">=</span><span class="s">&#39;CDS&#39;</span><span class="p">)]</span>
<span class="go">[&#39;CDS_1&#39;, &#39;CDS_2&#39;, &#39;CDS_3&#39;, &#39;CDS_4&#39;, &#39;CDS_5&#39;]</span>
</pre></div>
</div>
<p>Or as &#8220;grandchildren&#8221; of the gene:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">f</span><span class="o">.</span><span class="n">id</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">db</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="s">&#39;NC_000083.5:LOC100040603&#39;</span><span class="p">,</span> <span class="n">featuretype</span><span class="o">=</span><span class="s">&#39;CDS&#39;</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="mi">2</span><span class="p">)]</span>
<span class="go">[&#39;CDS_1&#39;, &#39;CDS_2&#39;, &#39;CDS_3&#39;, &#39;CDS_4&#39;, &#39;CDS_5&#39;]</span>
</pre></div>
</div>
<p>Note that the database id values are all unique for the CDSs, but their
<code class="docutils literal"><span class="pre">&quot;Name&quot;</span></code> attributes are still all the same, as expected:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">([</span><span class="n">f</span><span class="p">[</span><span class="s">&#39;Name&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">db</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="s">&#39;XM_001475631.1&#39;</span><span class="p">,</span> <span class="n">featuretype</span><span class="o">=</span><span class="s">&#39;CDS&#39;</span><span class="p">)]))</span>
<span class="go">[&#39;CDS:NC_000083.5:LOC100040603&#39;]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="c-elegans-ws199-ann-gff-txt">
<span id="id3"></span><h2>c_elegans_WS199_ann_gff.txt<a class="headerlink" href="#c-elegans-ws199-ann-gff-txt" title="Permalink to this headline">Â¶</a></h2>
<p>This is a minimal file, with only one feature and that one feature has no
genomic coordinates.  It illustrates how things behave when there is very
little information for a feature.</p>
<div class="section" id="id4">
<h3>File contents<a class="headerlink" href="#id4" title="Permalink to this headline">Â¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre># modified GFF file to remove location coordinates and test annotations
I	Expr_profile	experimental_result_region	.	.	.	+	.	expr_profile=B0019.1
</pre></div>
</div>
</div>
<div class="section" id="id5">
<h3>Import<a class="headerlink" href="#id5" title="Permalink to this headline">Â¶</a></h3>
<p>Import is straightforward:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fn</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span><span class="s">&#39;c_elegans_WS199_ann_gff.txt&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">create_db</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="s">&#39;:memory:&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="access">
<h3>Access<a class="headerlink" href="#access" title="Permalink to this headline">Â¶</a></h3>
<p>Confirm there&#8217;s only a single feature imported:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">db</span><span class="o">.</span><span class="n">all_features</span><span class="p">())</span>
<span class="go">[&lt;Feature experimental_result_region (I:.-.[+]) at 0x...]</span>
</pre></div>
</div>
<p>With no obvious ID fields in the attributes, we can access the single feature
by name using the feature type:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">db</span><span class="p">[</span><span class="s">&#39;experimental_result_region_1&#39;</span><span class="p">]</span>
<span class="go">&lt;Feature experimental_result_region (I:.-.[+]) at 0x...&gt;</span>
</pre></div>
</div>
<p>This shows that the internal representation of null values is None:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">f</span> <span class="o">=</span> <span class="n">db</span><span class="p">[</span><span class="s">&#39;experimental_result_region_1&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="n">f</span><span class="o">.</span><span class="n">start</span> <span class="ow">is</span> <span class="bp">None</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="n">f</span><span class="o">.</span><span class="n">stop</span> <span class="ow">is</span> <span class="bp">None</span>
</pre></div>
</div>
<p>But the string representation shows &#8220;<code class="docutils literal"><span class="pre">.</span></code>&#8221; placeholders:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">f</span><span class="p">)</span> 
<span class="go">I       Expr_profile    experimental_result_region      .       .       .       +       .       expr_profile=B0019.1</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="c-elegans-ws199-shortened-gff-txt">
<span id="id6"></span><h2>c_elegans_WS199_shortened_gff.txt<a class="headerlink" href="#c-elegans-ws199-shortened-gff-txt" title="Permalink to this headline">Â¶</a></h2>
<p>This file contains many different kinds of features; much like
<a class="reference internal" href="#mouse-extra-comma-gff3"><span>mouse_extra_comma.gff3</span></a> above, some features have IDs, some do not, and
there are many that share the same feature ID.  This example shows a common
way of subsetting features (here, getting all the non-historical CDSs for
a gene).</p>
<div class="section" id="id7">
<h3>File contents<a class="headerlink" href="#id7" title="Permalink to this headline">Â¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>I	Orfeome	PCR_product	12759747	12764936	.	-	.	amplified=1;pcr_product=mv_B0019.1
I	SAGE_tag_unambiguously_mapped	SAGE_tag	12763533	12763553	.	-	.	count=1;gene=amx-2;sequence=SAGE:ggcagagtcttttggca;transcript=B0019.1
I	SAGE_tag_unambiguously_mapped	SAGE_tag	12761492	12761512	.	-	.	count=5;gene=amx-2;sequence=SAGE:aacggagccgtacacgc;transcript=B0019.1
I	SAGE_tag_most_three_prime	SAGE_tag	12761499	12761512	.	-	.	count=9;gene=amx-2;sequence=SAGE:aacggagccg;transcript=B0019.1
X	SAGE_tag	SAGE_tag	6819353	6819366	.	+	.	count=9;gene=amx-2;sequence=SAGE:aacggagccg;transcript=B0019.1
I	Expr_profile	experimental_result_region	12762449	12764118	.	+	.	expr_profile=B0019.1
I	Coding_transcript	CDS	12759745	12759828	.	-	0	ID=CDS:B0019.1;Parent=Transcript:B0019.1;locus=amx-2;status=Partially_confirmed;wormpep=CE:CE40797
I	Coding_transcript	CDS	12759949	12760013	.	-	2	ID=CDS:B0019.1;Parent=Transcript:B0019.1;locus=amx-2;status=Partially_confirmed;wormpep=CE:CE40797
I	Coding_transcript	CDS	12760227	12760319	.	-	2	ID=CDS:B0019.1;Parent=Transcript:B0019.1;locus=amx-2;status=Partially_confirmed;wormpep=CE:CE40797
I	Coding_transcript	CDS	12760365	12760494	.	-	0	ID=CDS:B0019.1;Parent=Transcript:B0019.1;locus=amx-2;status=Partially_confirmed;wormpep=CE:CE40797
I	Coding_transcript	CDS	12760834	12760904	.	-	2	ID=CDS:B0019.1;Parent=Transcript:B0019.1;locus=amx-2;status=Partially_confirmed;wormpep=CE:CE40797
I	Coding_transcript	CDS	12761172	12761516	.	-	2	ID=CDS:B0019.1;Parent=Transcript:B0019.1;locus=amx-2;status=Partially_confirmed;wormpep=CE:CE40797
I	Coding_transcript	CDS	12761799	12761953	.	-	1	ID=CDS:B0019.1;Parent=Transcript:B0019.1;locus=amx-2;status=Partially_confirmed;wormpep=CE:CE40797
I	Coding_transcript	CDS	12762127	12762268	.	-	2	ID=CDS:B0019.1;Parent=Transcript:B0019.1;locus=amx-2;status=Partially_confirmed;wormpep=CE:CE40797
I	Coding_transcript	CDS	12762648	12762806	.	-	2	ID=CDS:B0019.1;Parent=Transcript:B0019.1;locus=amx-2;status=Partially_confirmed;wormpep=CE:CE40797
I	Coding_transcript	CDS	12763112	12763249	.	-	2	ID=CDS:B0019.1;Parent=Transcript:B0019.1;locus=amx-2;status=Partially_confirmed;wormpep=CE:CE40797
I	Coding_transcript	CDS	12763448	12763655	.	-	0	ID=CDS:B0019.1;Parent=Transcript:B0019.1;locus=amx-2;status=Partially_confirmed;wormpep=CE:CE40797
I	Coding_transcript	CDS	12763729	12763882	.	-	1	ID=CDS:B0019.1;Parent=Transcript:B0019.1;locus=amx-2;status=Partially_confirmed;wormpep=CE:CE40797
I	Coding_transcript	CDS	12763979	12764102	.	-	2	ID=CDS:B0019.1;Parent=Transcript:B0019.1;locus=amx-2;status=Partially_confirmed;wormpep=CE:CE40797
I	Coding_transcript	CDS	12764291	12764471	.	-	0	ID=CDS:B0019.1;Parent=Transcript:B0019.1;locus=amx-2;status=Partially_confirmed;wormpep=CE:CE40797
I	Coding_transcript	CDS	12764812	12764937	.	-	0	ID=CDS:B0019.1;Parent=Transcript:B0019.1;locus=amx-2;status=Partially_confirmed;wormpep=CE:CE40797
I	history	CDS	12759745	12759828	.	-	0	ID=CDS:B0019.1:wp173
I	history	CDS	12759949	12760013	.	-	2	ID=CDS:B0019.1:wp173
I	history	CDS	12760227	12760319	.	-	2	ID=CDS:B0019.1:wp173
I	history	CDS	12760365	12760494	.	-	0	ID=CDS:B0019.1:wp173
I	history	CDS	12760834	12760904	.	-	2	ID=CDS:B0019.1:wp173
I	history	CDS	12761172	12761516	.	-	2	ID=CDS:B0019.1:wp173
I	history	CDS	12761577	12761626	.	-	1	ID=CDS:B0019.1:wp173
I	history	CDS	12761795	12761953	.	-	1	ID=CDS:B0019.1:wp173
I	history	CDS	12762127	12762268	.	-	2	ID=CDS:B0019.1:wp173
I	history	CDS	12762648	12762806	.	-	2	ID=CDS:B0019.1:wp173
I	history	CDS	12763112	12763249	.	-	2	ID=CDS:B0019.1:wp173
I	history	CDS	12763448	12763655	.	-	0	ID=CDS:B0019.1:wp173
I	history	CDS	12763729	12763882	.	-	1	ID=CDS:B0019.1:wp173
I	history	CDS	12763979	12764102	.	-	2	ID=CDS:B0019.1:wp173
I	history	CDS	12764291	12764471	.	-	0	ID=CDS:B0019.1:wp173
I	history	CDS	12764812	12764937	.	-	0	ID=CDS:B0019.1:wp173
I	history	CDS	12759745	12759828	.	-	0	ID=CDS:B0019.1:wp90
I	history	CDS	12759949	12760013	.	-	2	ID=CDS:B0019.1:wp90
I	history	CDS	12760227	12760319	.	-	2	ID=CDS:B0019.1:wp90
I	history	CDS	12761172	12761516	.	-	2	ID=CDS:B0019.1:wp90
I	history	CDS	12761577	12761626	.	-	1	ID=CDS:B0019.1:wp90
I	history	CDS	12761795	12761953	.	-	1	ID=CDS:B0019.1:wp90
I	history	CDS	12762127	12762268	.	-	2	ID=CDS:B0019.1:wp90
I	history	CDS	12762648	12762806	.	-	2	ID=CDS:B0019.1:wp90
I	history	CDS	12763112	12763249	.	-	2	ID=CDS:B0019.1:wp90
I	history	CDS	12763469	12763655	.	-	0	ID=CDS:B0019.1:wp90
I	history	CDS	12763729	12763882	.	-	1	ID=CDS:B0019.1:wp90
I	history	CDS	12763979	12764102	.	-	2	ID=CDS:B0019.1:wp90
I	history	CDS	12764291	12764471	.	-	0	ID=CDS:B0019.1:wp90
I	history	CDS	12764812	12764937	.	-	0	ID=CDS:B0019.1:wp90
I	mass_spec_genome	translated_nucleotide_match	12761920	12761953	.	-	.	ID=Target:381130;Target=Mass_spec_peptide:MSP:FADFSPLDVSDVNFATDDLAK 10 21 +;Note=MSP:FADFSPLDVSDVNFATDDLAK;cds_matches=B0019.1;protein_matches=WP:CE40797;times_observed=3
I	mass_spec_genome	translated_nucleotide_match	12762127	12762155	.	-	.	ID=Target:381130;Target=Mass_spec_peptide:MSP:FADFSPLDVSDVNFATDDLAK 1 10 +;Note=MSP:FADFSPLDVSDVNFATDDLAK;cds_matches=B0019.1;protein_matches=WP:CE40797;times_observed=3
I	mass_spec_genome	translated_nucleotide_match	12763506	12763559	.	-	.	ID=Target:381133;Target=Mass_spec_peptide:MSP:FGHGQSLLAQGGMNEVVR 1 18 +;Note=MSP:FGHGQSLLAQGGMNEVVR;cds_matches=B0019.1;protein_matches=WP:CE40797;times_observed=1
I	mass_spec_genome	translated_nucleotide_match	12764361	12764411	.	-	.	ID=Target:381144;Target=Mass_spec_peptide:MSP:NIQQNRPGLSVLVLEAR 1 17 +;Note=MSP:NIQQNRPGLSVLVLEAR;cds_matches=B0019.1;protein_matches=WP:CE40797;times_observed=2
I	Coding_transcript	mRNA	12759582	12764949	.	-	.	ID=Transcript:B0019.1;Note=amx-2;Parent=Gene:WBGene00000138;cds=B0019.1;prediction_status=Partially_confirmed;wormpep=CE:CE40797
I	Allele	SNP	12764272	12764272	.	+	.	interpolated_map_position=14.003;rflp=No;variation=snp_B0019[1]
I	Oligo_set	reagent	12759745	12761589	.	-	.	oligo_set=Aff_B0019.1
I	Coding_transcript	exon	12759745	12759828	.	-	0	Parent=Transcript:B0019.1
I	Coding_transcript	exon	12759949	12760013	.	-	2	Parent=Transcript:B0019.1
I	Coding_transcript	exon	12760227	12760319	.	-	2	Parent=Transcript:B0019.1
I	Coding_transcript	exon	12760365	12760494	.	-	0	Parent=Transcript:B0019.1
I	Coding_transcript	exon	12760834	12760904	.	-	2	Parent=Transcript:B0019.1
I	Coding_transcript	exon	12761172	12761516	.	-	2	Parent=Transcript:B0019.1
I	Coding_transcript	exon	12761799	12761953	.	-	1	Parent=Transcript:B0019.1
I	Coding_transcript	exon	12762127	12762268	.	-	2	Parent=Transcript:B0019.1
I	Coding_transcript	exon	12762648	12762806	.	-	2	Parent=Transcript:B0019.1
I	Coding_transcript	exon	12763112	12763249	.	-	2	Parent=Transcript:B0019.1
I	Coding_transcript	exon	12763448	12763655	.	-	0	Parent=Transcript:B0019.1
I	Coding_transcript	exon	12763729	12763882	.	-	1	Parent=Transcript:B0019.1
I	Coding_transcript	exon	12763979	12764102	.	-	2	Parent=Transcript:B0019.1
I	Coding_transcript	exon	12764291	12764471	.	-	0	Parent=Transcript:B0019.1
I	Coding_transcript	exon	12764812	12764937	.	-	0	Parent=Transcript:B0019.1
I	Coding_transcript	five_prime_UTR	12764938	12764949	.	-	.	Parent=Transcript:B0019.1
I	Coding_transcript	three_prime_UTR	12759582	12759744	.	-	.	Parent=Transcript:B0019.1
I	Coding_transcript	intron	12760495	12760833	.	-	.	Parent=Transcript:B0019.1;confirmed_est=EC027594
I	Coding_transcript	intron	12760905	12761171	.	-	.	Parent=Transcript:B0019.1;confirmed_est=EC027594
I	Coding_transcript	intron	12761517	12761798	.	-	.	Parent=Transcript:B0019.1;confirmed_est=EC027594
I	Coding_transcript	intron	12759829	12759948	.	-	.	Parent=Transcript:B0019.1;confirmed_est=EC034652
I	Coding_transcript	intron	12760014	12760226	.	-	.	Parent=Transcript:B0019.1;confirmed_est=EC034652
I	Coding_transcript	intron	12760320	12760364	.	-	.	Parent=Transcript:B0019.1;confirmed_est=yk1054h04.3
I	Coding_transcript	intron	12763883	12763978	.	-	.	Parent=Transcript:B0019.1;confirmed_est=yk1054h04.5,OSTF088D9_1
I	Coding_transcript	intron	12764103	12764290	.	-	.	Parent=Transcript:B0019.1;confirmed_est=yk1054h04.5,OSTF088D9_1
I	Coding_transcript	intron	12764472	12764811	.	-	.	Parent=Transcript:B0019.1;confirmed_est=yk1054h04.5,OSTF088D9_1
I	Coding_transcript	intron	12762807	12763111	.	-	.	Parent=Transcript:B0019.1;confirmed_est=yk1056c07.5
I	Coding_transcript	intron	12763250	12763447	.	-	.	Parent=Transcript:B0019.1;confirmed_est=yk1056c07.5
I	Coding_transcript	intron	12763656	12763728	.	-	.	Parent=Transcript:B0019.1;confirmed_est=yk1056c07.5
I	Coding_transcript	intron	12761954	12762126	.	-	.	Parent=Transcript:B0019.1;confirmed_est=yk262g9.5
I	Coding_transcript	intron	12762269	12762647	.	-	.	Parent=Transcript:B0019.1;confirmed_est=yk262g9.5
I	Promoterome	PCR_product	12764938	12766937	.	+	.	pcr_product=p_B0019.1_93
I	GenePair_STS	PCR_product	12762449	12764118	.	+	.	pcr_product=sjj_B0019.1
I	Coding_transcript	gene	12759582	12764949	.	-	.	ID=Gene:WBGene00000138
III	Orfeome	PCR_product	13780230	13780850	.	+	.	amplified=1;pcr_product=mv_3R5.1.v6
IV	Orfeome	PCR_product	17486939	17488952	.	-	.	amplified=1;pcr_product=mv_4R79.1
IV	Orfeome	PCR_product	17480353	17483284	.	-	.	amplified=1;pcr_product=mv_4R79.2
X	Orfeome	PCR_product	17714881	17718531	.	+	.	amplified=1;pcr_product=mv_6R55.1
X	Orfeome	PCR_product	17712787	17714742	.	+	.	amplified=1;pcr_product=mv_6R55.2
II	Orfeome	PCR_product	6995874	7010146	.	+	.	amplified=1;pcr_product=mv_AAA03517
III	Orfeome	PCR_product	5625097	5631795	.	+	.	amplified=1;pcr_product=mv_AAA03544
X	GenePair_STS	PCR_product	9962853	9963737	.	+	.	pcr_product=cenix:102-c3
II	GenePair_STS	PCR_product	5507236	5508135	.	+	.	pcr_product=cenix:102-c4
V	GenePair_STS	PCR_product	10117842	10118735	.	+	.	pcr_product=cenix:102-c5
IV	GenePair_STS	PCR_product	3566130	3567025	.	+	.	pcr_product=cenix:102-c6
X	GenePair_STS	PCR_product	6117180	6117930	.	+	.	pcr_product=cenix:102-c7
IV	GenePair_STS	PCR_product	7189492	7190369	.	+	.	pcr_product=cenix:102-c9
II	GenePair_STS	PCR_product	14462527	14463202	.	+	.	pcr_product=cenix:102-d1
X	Promoterome	PCR_product	2258069	2259336	.	+	.	pcr_product=p_AH9.2_93
IV	Promoterome	PCR_product	12157449	12159448	.	+	.	pcr_product=p_B0001.6_93
I	Promoterome	PCR_product	12764938	12766937	.	+	.	pcr_product=p_B0019.1_93
V	Promoterome	PCR_product	10320122	10320689	.	+	.	pcr_product=p_B0024.12_93
I	Coding_transcript	CDS	4581214	4581237	.	-	0	ID=CDS:D1007.5b;Parent=Transcript:D1007.5b.2,Transcript:D1007.5b.1;status=Confirmed;wormpep=WP:CE33577
I	Coding_transcript	CDS	4581664	4582026	.	-	0	ID=CDS:D1007.5b;Parent=Transcript:D1007.5b.2,Transcript:D1007.5b.1;status=Confirmed;wormpep=WP:CE33577
I	Coding_transcript	CDS	4582412	4582718	.	-	1	ID=CDS:D1007.5b;Parent=Transcript:D1007.5b.2,Transcript:D1007.5b.1;status=Confirmed;wormpep=WP:CE33577
I	Coding_transcript	CDS	4583190	4583374	.	-	0	ID=CDS:D1007.5b;Parent=Transcript:D1007.5b.2,Transcript:D1007.5b.1;status=Confirmed;wormpep=WP:CE33577
I	Coding_transcript	CDS	4583426	4583509	.	-	0	ID=CDS:D1007.5b;Parent=Transcript:D1007.5b.2,Transcript:D1007.5b.1;status=Confirmed;wormpep=WP:CE33577
I	Coding_transcript	CDS	4583560	4583805	.	-	0	ID=CDS:D1007.5b;Parent=Transcript:D1007.5b.2,Transcript:D1007.5b.1;status=Confirmed;wormpep=WP:CE33577
I	Coding_transcript	mRNA	4580734	4583815	.	-	.	ID=Transcript:D1007.5b.1;Parent=Gene:WBGene00017003;cds=D1007.5b;prediction_status=Confirmed;wormpep=WP:CE33577
I	Coding_transcript	mRNA	4581214	4583811	.	-	.	ID=Transcript:D1007.5b.2;Parent=Gene:WBGene00017003;cds=D1007.5b;prediction_status=Confirmed;wormpep=WP:CE33577
I	Coding_transcript	exon	4581214	4581237	.	-	0	Parent=Transcript:D1007.5b.1
I	Coding_transcript	exon	4581664	4582026	.	-	0	Parent=Transcript:D1007.5b.1
I	Coding_transcript	exon	4582412	4582718	.	-	1	Parent=Transcript:D1007.5b.1
I	Coding_transcript	exon	4583190	4583374	.	-	0	Parent=Transcript:D1007.5b.1
I	Coding_transcript	exon	4583426	4583509	.	-	0	Parent=Transcript:D1007.5b.1
I	Coding_transcript	exon	4583560	4583805	.	-	0	Parent=Transcript:D1007.5b.1
I	Coding_transcript	five_prime_UTR	4583806	4583815	.	-	.	Parent=Transcript:D1007.5b.1
I	Coding_transcript	three_prime_UTR	4580734	4581213	.	-	.	Parent=Transcript:D1007.5b.1
I	Coding_transcript	intron	4582027	4582411	.	-	.	Parent=Transcript:D1007.5b.1;confirmed_est=EB994038
I	Coding_transcript	intron	4583375	4583425	.	-	.	Parent=Transcript:D1007.5b.1;confirmed_est=EC038345,OSTF085G5_1
I	Coding_transcript	intron	4583510	4583559	.	-	.	Parent=Transcript:D1007.5b.1;confirmed_est=EC038345,OSTF085G5_1
I	Coding_transcript	intron	4582719	4583189	.	-	.	Parent=Transcript:D1007.5b.1;confirmed_est=yk1055g06.5,OSTF085G5_1
I	Coding_transcript	intron	4581238	4581663	.	-	.	Parent=Transcript:D1007.5b.1;confirmed_est=yk1057e08.3
I	Coding_transcript	exon	4581214	4581237	.	-	0	Parent=Transcript:D1007.5b.2
I	Coding_transcript	exon	4581664	4582026	.	-	0	Parent=Transcript:D1007.5b.2
I	Coding_transcript	exon	4582412	4582718	.	-	1	Parent=Transcript:D1007.5b.2
I	Coding_transcript	exon	4583190	4583374	.	-	0	Parent=Transcript:D1007.5b.2
I	Coding_transcript	exon	4583426	4583509	.	-	0	Parent=Transcript:D1007.5b.2
I	Coding_transcript	exon	4583560	4583805	.	-	0	Parent=Transcript:D1007.5b.2
I	Coding_transcript	five_prime_UTR	4583806	4583811	.	-	.	Parent=Transcript:D1007.5b.2
I	Coding_transcript	intron	4582027	4582411	.	-	.	Parent=Transcript:D1007.5b.2;confirmed_est=EB994038
I	Coding_transcript	intron	4583375	4583425	.	-	.	Parent=Transcript:D1007.5b.2;confirmed_est=EC038345,OSTF085G5_1
I	Coding_transcript	intron	4583510	4583559	.	-	.	Parent=Transcript:D1007.5b.2;confirmed_est=EC038345,OSTF085G5_1
I	Coding_transcript	intron	4582719	4583189	.	-	.	Parent=Transcript:D1007.5b.2;confirmed_est=yk1055g06.5,OSTF085G5_1
I	Coding_transcript	intron	4581238	4581663	.	-	.	Parent=Transcript:D1007.5b.2;confirmed_est=yk1057e08.3
I	Coding_transcript	gene	4580693	4583815	.	-	.	ID=Gene:WBGene00017003
I	SAGE_tag_unambiguously_mapped	SAGE_tag	4581093	4581113	.	-	.	count=10;gene=D1007.5;sequence=SAGE:tttgcgaattacttgct;transcript=D1007.5b.1,D1007.5a
I	SAGE_tag_unambiguously_mapped	SAGE_tag	4580748	4580768	.	-	.	count=112;gene=D1007.5;sequence=SAGE:ttttccattaattttga;transcript=D1007.5b.1,D1007.5a
I	SAGE_tag_unambiguously_mapped	SAGE_tag	4582415	4582428	.	-	.	count=1;gene=D1007.5;sequence=SAGE:cattttcgtg;transcript=D1007.5b.2,D1007.5b.1,D1007.5a
I	SAGE_tag_unambiguously_mapped	SAGE_tag	4580914	4580927	.	-	.	count=1;gene=D1007.5;sequence=SAGE:taaatttcaa;transcript=D1007.5b.1,D1007.5a
I	SAGE_tag_unambiguously_mapped	SAGE_tag	4581193	4581206	.	-	.	count=1;gene=D1007.5;sequence=SAGE:tgctcgttcg;transcript=D1007.5b.1,D1007.5a
I	SAGE_tag_unambiguously_mapped	SAGE_tag	4583465	4583478	.	-	.	count=1;gene=D1007.5;sequence=SAGE:tgttggcctt;transcript=D1007.5b.2,D1007.5b.1,D1007.5a
I	SAGE_tag_unambiguously_mapped	SAGE_tag	4583458	4583478	.	-	.	count=1;gene=D1007.5;sequence=SAGE:tgttggccttttacttg;transcript=D1007.5b.2,D1007.5b.1,D1007.5a
I	SAGE_tag_unambiguously_mapped	SAGE_tag	4582533	4582553	.	-	.	count=2;gene=D1007.5;sequence=SAGE:tgcagtgatagtccagc;transcript=D1007.5b.2,D1007.5b.1,D1007.5a
I	SAGE_tag_unambiguously_mapped	SAGE_tag	4581100	4581113	.	-	.	count=2;gene=D1007.5;sequence=SAGE:tttgcgaatt;transcript=D1007.5b.1,D1007.5a
I	SAGE_tag_unambiguously_mapped	SAGE_tag	4580755	4580768	.	-	.	count=43;gene=D1007.5;sequence=SAGE:ttttccatta;transcript=D1007.5b.1,D1007.5a
I	Coding_transcript	CDS	4580993	4581241	.	-	0	ID=CDS:D1007.5a;Parent=Transcript:D1007.5a;status=Confirmed;wormpep=CE:CE29034
I	Coding_transcript	CDS	4581664	4582026	.	-	0	ID=CDS:D1007.5a;Parent=Transcript:D1007.5a;status=Confirmed;wormpep=CE:CE29034
I	Coding_transcript	CDS	4582412	4582718	.	-	1	ID=CDS:D1007.5a;Parent=Transcript:D1007.5a;status=Confirmed;wormpep=CE:CE29034
I	Coding_transcript	CDS	4583190	4583374	.	-	0	ID=CDS:D1007.5a;Parent=Transcript:D1007.5a;status=Confirmed;wormpep=CE:CE29034
I	Coding_transcript	CDS	4583426	4583509	.	-	0	ID=CDS:D1007.5a;Parent=Transcript:D1007.5a;status=Confirmed;wormpep=CE:CE29034
I	Coding_transcript	CDS	4583560	4583805	.	-	0	ID=CDS:D1007.5a;Parent=Transcript:D1007.5a;status=Confirmed;wormpep=CE:CE29034
I	mass_spec_genome	translated_nucleotide_match	4580996	4581052	.	-	.	ID=Target:277116;Target=Mass_spec_peptide:MSP:IYEPSQEDLLLMHQLQQER 1 19 +;Note=MSP:IYEPSQEDLLLMHQLQQER;cds_matches=D1007.5a;protein_matches=WP:CE29034;times_observed=1
I	mass_spec_genome	translated_nucleotide_match	4581838	4581882	.	-	.	ID=Target:277138;Target=Mass_spec_peptide:MSP:AAIHLGSWHQIEGPR 1 15 +;Note=MSP:AAIHLGSWHQIEGPR;cds_matches=D1007.5b D1007.5a;protein_matches=WP:CE33577 WP:CE29034;times_observed=1
I	mass_spec_genome	translated_nucleotide_match	4583581	4583601	.	-	.	ID=Target:277176;Target=Mass_spec_peptide:MSP:TLWWLPK 1 7 +;Note=MSP:TLWWLPK;cds_matches=D1007.5b D1007.5a;protein_matches=WP:CE33577 WP:CE29034;times_observed=1
I	Coding_transcript	mRNA	4580693	4583811	.	-	.	ID=Transcript:D1007.5a;Parent=Gene:WBGene00017003;cds=D1007.5a;prediction_status=Confirmed;wormpep=CE:CE29034
I	Coding_transcript	exon	4580993	4581241	.	-	0	Parent=Transcript:D1007.5a
I	Coding_transcript	exon	4581664	4582026	.	-	0	Parent=Transcript:D1007.5a
I	Coding_transcript	exon	4582412	4582718	.	-	1	Parent=Transcript:D1007.5a
I	Coding_transcript	exon	4583190	4583374	.	-	0	Parent=Transcript:D1007.5a
I	Coding_transcript	exon	4583426	4583509	.	-	0	Parent=Transcript:D1007.5a
I	Coding_transcript	exon	4583560	4583805	.	-	0	Parent=Transcript:D1007.5a
I	Coding_transcript	five_prime_UTR	4583806	4583811	.	-	.	Parent=Transcript:D1007.5a
I	Coding_transcript	three_prime_UTR	4580693	4580992	.	-	.	Parent=Transcript:D1007.5a
I	Coding_transcript	intron	4582027	4582411	.	-	.	Parent=Transcript:D1007.5a;confirmed_est=EB994038
I	Coding_transcript	intron	4581242	4581663	.	-	.	Parent=Transcript:D1007.5a;confirmed_est=EB994038,OSTR085G5_1
I	Coding_transcript	intron	4583375	4583425	.	-	.	Parent=Transcript:D1007.5a;confirmed_est=EC038345,OSTF085G5_1
I	Coding_transcript	intron	4583510	4583559	.	-	.	Parent=Transcript:D1007.5a;confirmed_est=EC038345,OSTF085G5_1
I	Coding_transcript	intron	4582719	4583189	.	-	.	Parent=Transcript:D1007.5a;confirmed_est=yk1055g06.5,OSTF085G5_1
</pre></div>
</div>
</div>
<div class="section" id="id8">
<h3>Import<a class="headerlink" href="#id8" title="Permalink to this headline">Â¶</a></h3>
<p>If <code class="docutils literal"><span class="pre">merge_strategy=&quot;merge&quot;</span></code>, is used here, exceptions will be raised because
all those CDSs cannot be logically merged: they have different
coordinates. So <code class="docutils literal"><span class="pre">merge_strategy=&quot;create_unique&quot;</span></code> is used instead. This kwarg
will add an underscore and an integer to each duplicate value used as a primary
key:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fn</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span><span class="s">&#39;c_elegans_WS199_shortened_gff.txt&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">create_db</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="s">&quot;:memory:&quot;</span><span class="p">,</span> <span class="n">merge_strategy</span><span class="o">=</span><span class="s">&#39;create_unique&#39;</span><span class="p">,</span> <span class="n">keep_order</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="id9">
<h3>Access<a class="headerlink" href="#id9" title="Permalink to this headline">Â¶</a></h3>
<p>Getting a gene:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">db</span><span class="p">[</span><span class="s">&#39;Gene:WBGene00017003&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span>
<span class="go">&lt;Feature gene (I:4580693-4583815[-]) at 0x...&gt;</span>
</pre></div>
</div>
<p>Get all &#8220;non-historical&#8221; CDSs for this gene.  This illustrates that:</p>
<ol class="arabic simple">
<li>you can use a <code class="xref py py-class docutils literal"><span class="pre">Feature</span></code> object instead of a string for the database primary key</li>
<li>printing a <code class="xref py py-class docutils literal"><span class="pre">Feature</span></code> reconstructs the line, and</li>
<li>the string representation of a line does not include a <code class="docutils literal"><span class="pre">\n</span></code>.</li>
</ol>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">db</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">featuretype</span><span class="o">=</span><span class="s">&#39;CDS&#39;</span><span class="p">,</span> <span class="n">order_by</span><span class="o">=</span><span class="s">&#39;start&#39;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">i</span><span class="o">.</span><span class="n">source</span> <span class="o">!=</span> <span class="s">&#39;history&#39;</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">print</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>  
<span class="go">I       Coding_transcript       CDS     4580993 4581241 .       -       0       ID=CDS:D1007.5a;Parent=Transcript:D1007.5a;status=Confirmed;wormpep=CE:CE29034</span>
<span class="go">I       Coding_transcript       CDS     4581214 4581237 .       -       0       ID=CDS:D1007.5b;Parent=Transcript:D1007.5b.2,Transcript:D1007.5b.1;status=Confirmed;wormpep=WP:CE33577</span>
<span class="go">I       Coding_transcript       CDS     4581664 4582026 .       -       0       ID=CDS:D1007.5a;Parent=Transcript:D1007.5a;status=Confirmed;wormpep=CE:CE29034</span>
<span class="go">I       Coding_transcript       CDS     4581664 4582026 .       -       0       ID=CDS:D1007.5b;Parent=Transcript:D1007.5b.2,Transcript:D1007.5b.1;status=Confirmed;wormpep=WP:CE33577</span>
<span class="go">I       Coding_transcript       CDS     4582412 4582718 .       -       1       ID=CDS:D1007.5a;Parent=Transcript:D1007.5a;status=Confirmed;wormpep=CE:CE29034</span>
<span class="go">I       Coding_transcript       CDS     4582412 4582718 .       -       1       ID=CDS:D1007.5b;Parent=Transcript:D1007.5b.2,Transcript:D1007.5b.1;status=Confirmed;wormpep=WP:CE33577</span>
<span class="go">I       Coding_transcript       CDS     4583190 4583374 .       -       0       ID=CDS:D1007.5a;Parent=Transcript:D1007.5a;status=Confirmed;wormpep=CE:CE29034</span>
<span class="go">I       Coding_transcript       CDS     4583190 4583374 .       -       0       ID=CDS:D1007.5b;Parent=Transcript:D1007.5b.2,Transcript:D1007.5b.1;status=Confirmed;wormpep=WP:CE33577</span>
<span class="go">I       Coding_transcript       CDS     4583426 4583509 .       -       0       ID=CDS:D1007.5a;Parent=Transcript:D1007.5a;status=Confirmed;wormpep=CE:CE29034</span>
<span class="go">I       Coding_transcript       CDS     4583426 4583509 .       -       0       ID=CDS:D1007.5b;Parent=Transcript:D1007.5b.2,Transcript:D1007.5b.1;status=Confirmed;wormpep=WP:CE33577</span>
<span class="go">I       Coding_transcript       CDS     4583560 4583805 .       -       0       ID=CDS:D1007.5a;Parent=Transcript:D1007.5a;status=Confirmed;wormpep=CE:CE29034</span>
<span class="go">I       Coding_transcript       CDS     4583560 4583805 .       -       0       ID=CDS:D1007.5b;Parent=Transcript:D1007.5b.2,Transcript:D1007.5b.1;status=Confirmed;wormpep=WP:CE33577</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="ensembl-gtf-txt">
<span id="id10"></span><h2>ensembl_gtf.txt<a class="headerlink" href="#ensembl-gtf-txt" title="Permalink to this headline">Â¶</a></h2>
<p>In this GTF file, gene_id and transcript_id are identical, creating problems
for unique IDs in the database.</p>
<div class="section" id="id11">
<h3>File contents<a class="headerlink" href="#id11" title="Permalink to this headline">Â¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>I	snoRNA	exon	3747	3909	.	-	.	 gene_id &quot;Y74C9A.6&quot;; transcript_id &quot;Y74C9A.6&quot;; exon_number &quot;1&quot;; gene_name &quot;Y74C9A.6&quot;; transcript_name &quot;NR_001477.2&quot;;
I	protein_coding	exon	12764812	12764949	.	-	.	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;1&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	CDS	12764812	12764937	.	-	0	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;1&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;; protein_id &quot;B0019.1&quot;;
I	protein_coding	start_codon	12764935	12764937	.	-	0	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;1&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	exon	12764291	12764471	.	-	.	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;2&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	CDS	12764291	12764471	.	-	0	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;2&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;; protein_id &quot;B0019.1&quot;;
I	protein_coding	exon	12763979	12764102	.	-	.	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;3&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	CDS	12763979	12764102	.	-	2	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;3&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;; protein_id &quot;B0019.1&quot;;
I	protein_coding	exon	12763729	12763882	.	-	.	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;4&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	CDS	12763729	12763882	.	-	1	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;4&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;; protein_id &quot;B0019.1&quot;;
I	protein_coding	exon	12763448	12763655	.	-	.	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;5&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	CDS	12763448	12763655	.	-	0	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;5&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;; protein_id &quot;B0019.1&quot;;
I	protein_coding	exon	12763112	12763249	.	-	.	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;6&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	CDS	12763112	12763249	.	-	2	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;6&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;; protein_id &quot;B0019.1&quot;;
I	protein_coding	exon	12762648	12762806	.	-	.	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;7&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	CDS	12762648	12762806	.	-	2	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;7&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;; protein_id &quot;B0019.1&quot;;
I	protein_coding	exon	12762127	12762268	.	-	.	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;8&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	CDS	12762127	12762268	.	-	2	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;8&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;; protein_id &quot;B0019.1&quot;;
I	protein_coding	exon	12761799	12761953	.	-	.	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;9&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	CDS	12761799	12761953	.	-	1	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;9&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;; protein_id &quot;B0019.1&quot;;
I	protein_coding	exon	12761172	12761516	.	-	.	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;10&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	CDS	12761172	12761516	.	-	2	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;10&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;; protein_id &quot;B0019.1&quot;;
I	protein_coding	exon	12760834	12760904	.	-	.	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;11&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	CDS	12760834	12760904	.	-	2	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;11&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;; protein_id &quot;B0019.1&quot;;
I	protein_coding	exon	12760365	12760494	.	-	.	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;12&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	CDS	12760365	12760494	.	-	0	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;12&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;; protein_id &quot;B0019.1&quot;;
I	protein_coding	exon	12760227	12760319	.	-	.	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;13&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	CDS	12760227	12760319	.	-	2	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;13&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;; protein_id &quot;B0019.1&quot;;
I	protein_coding	exon	12759949	12760013	.	-	.	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;14&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	CDS	12759949	12760013	.	-	2	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;14&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;; protein_id &quot;B0019.1&quot;;
I	protein_coding	exon	12759579	12759828	.	-	.	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;15&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
I	protein_coding	CDS	12759748	12759828	.	-	0	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;15&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;; protein_id &quot;B0019.1&quot;;
I	protein_coding	stop_codon	12759745	12759747	.	-	0	 gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1&quot;; exon_number &quot;15&quot;; gene_name &quot;amx-2&quot;; transcript_name &quot;B0019.1&quot;;
</pre></div>
</div>
</div>
<div class="section" id="id12">
<h3>Import<a class="headerlink" href="#id12" title="Permalink to this headline">Â¶</a></h3>
<p>As part of the import process, <a class="reference internal" href="api.html#module-gffutils" title="gffutils"><code class="xref py py-mod docutils literal"><span class="pre">gffutils</span></code></a> creates new &#8220;derived&#8221; features
for genes and transcripts (see <a class="reference internal" href="gtf.html#gtf"><span>GTF files</span></a>).  These are not explicitly defined in a GTF file, but
can be inferred.  Even though gene and transcript features do not exist in the
file, they will still be created.  This is why <code class="docutils literal"><span class="pre">id_spec={'gene':</span> <span class="pre">'gene_id'}</span></code>
works: the primary key of &#8220;gene&#8221; features &#8211; which don&#8217;t exist in the GTF file
but are inferred by <a class="reference internal" href="api.html#module-gffutils" title="gffutils"><code class="xref py py-mod docutils literal"><span class="pre">gffutils</span></code></a> &#8211; will be the &#8220;gene_id&#8221; attribute.</p>
<p>Note that in this example, gene_id and transcript_id are identical.  This
causes all sorts of problems.  To get around this we can write a transform
function (see <a class="reference internal" href="database-ids.html#transform"><span>transform</span></a> for details) that applies an arbitrary
transformation to a dictionary returned by the parser.  This will modify the
data upon import into the database:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">transform_func</span><span class="p">(</span><span class="n">x</span><span class="p">):</span>
<span class="gp">... </span>    <span class="c"># adds some text to the end of transcript IDs</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="s">&#39;transcript_id&#39;</span> <span class="ow">in</span> <span class="n">x</span><span class="o">.</span><span class="n">attributes</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">x</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;transcript_id&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&#39;_transcript&#39;</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">x</span>
</pre></div>
</div>
<p>Now we can supply this tranform function to <code class="xref py py-func docutils literal"><span class="pre">create_db()</span></code>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fn</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span><span class="s">&#39;ensembl_gtf.txt&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">create_db</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="s">&quot;:memory:&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">id_spec</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;gene&#39;</span><span class="p">:</span> <span class="s">&#39;gene_id&#39;</span><span class="p">,</span> <span class="s">&#39;transcript&#39;</span><span class="p">:</span> <span class="s">&quot;transcript_id&quot;</span><span class="p">},</span>
<span class="gp">... </span><span class="n">merge_strategy</span><span class="o">=</span><span class="s">&quot;create_unique&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">transform</span><span class="o">=</span><span class="n">transform_func</span><span class="p">,</span>
<span class="gp">... </span><span class="n">keep_order</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="id13">
<h3>Access<a class="headerlink" href="#id13" title="Permalink to this headline">Â¶</a></h3>
<p>We can access the derived &#8220;gene&#8221; feature by its ID (recall it doesn&#8217;t actually
exist in the GTF file):</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span> <span class="o">=</span> <span class="n">db</span><span class="p">[</span><span class="s">&quot;B0019.1&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">g</span>
<span class="go">&lt;Feature gene (I:12759579-12764949[-]) at 0x...&gt;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">g</span><span class="p">)</span>  
<span class="go">I       gffutils_derived        gene    12759579        12764949        .       -       .       gene_id &quot;B0019.1&quot;;</span>
</pre></div>
</div>
<p>And the derived &#8220;transcript&#8221; by its ID &#8211; which now has &#8220;_transcript&#8221; on the
end because of that transform function:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">=</span> <span class="n">db</span><span class="p">[</span><span class="s">&quot;B0019.1_transcript&quot;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">t</span>
<span class="go">&lt;Feature transcript (I:12759579-12764949[-]) at 0x...&gt;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>  
<span class="go">I       gffutils_derived        transcript      12759579        12764949        .       -       .       gene_id &quot;B0019.1&quot;; transcript_id &quot;B0019.1_transcript&quot;;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="f3-unique-3-v2-gff">
<span id="id14"></span><h2>F3-unique-3.v2.gff<a class="headerlink" href="#f3-unique-3-v2-gff" title="Permalink to this headline">Â¶</a></h2>
<p>This file does not contain hierarchical information, and has many directives.
Since much of the benefit of <a class="reference internal" href="api.html#module-gffutils" title="gffutils"><code class="xref py py-mod docutils literal"><span class="pre">gffutils</span></code></a> comes from being able to navigate
the hierarchical relationships, it might be overkill to use <a class="reference internal" href="api.html#module-gffutils" title="gffutils"><code class="xref py py-mod docutils literal"><span class="pre">gffutils</span></code></a> on
this file.  An exeception would be if you want to be able to look up the read
names by their ID.  Here, we use yet a differnt way of overriding the way
<a class="reference internal" href="api.html#module-gffutils" title="gffutils"><code class="xref py py-mod docutils literal"><span class="pre">gffutils</span></code></a> decides on a primary key for the database.</p>
<div class="section" id="id15">
<h3>File contents<a class="headerlink" href="#id15" title="Permalink to this headline">Â¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>##solid-gff-version 0.2
##gff-version 2
##source-version MaToGff.java v1.5
##date 2008-05-28
##time 13:11:03
##Type solid_read
##color-code AA=0,AC=1,AG=2,AT=3,CA=1,CC=0,CG=3,CT=2,GA=2,GC=3,GG=0,GT=1,TA=3,TC=2,TG=1,TT=0
##primer-base F3=T
##max-num-mismatches 3
##max-read-length 20
##line-order fragment
##history filter_fasta.pl --noduplicates --output=/data/results/DAEMON/DAEMON_MATE_PAIRS_2_20070326/S1/results.01/primary.20071218094706805 --name=DAEMON_MATE_PAIRS_2_20070326_S1 --tag=F3 --minlength=20 --prefix=T /data/results/DAEMON/DAEMON_MATE_PAIRS_2_20070326/S1/jobs/postPrimerSetPrimary.117/rawseq
##history map /data/results/RegressionDriver/CaseManager/results/r12/integration/case0002/reads1/test_S1_F3.csfasta /data/results/RegressionDriver/CaseManager/knownData/validatedReference/matchingPipeline/ecoli_k12_MG1655.fasta T=30 L=19 C=1 E=.Tmpfile1211939575SVhDtd F=0 B=1 D=1 u=1 r=0 n=1 Z=1000 P=&quot;0000000111111111111&quot; M=0 U=0.000000 H=0  &gt; .Tmpfile1211939575SVhDtd.out.1
##history MaToGff.java --sort --qvs=test_S1_F3_QV.qual.txt --convert=unique --clear=3 --tempdir=../tmp test_S1_F3.csfasta.ma.20.3
##hdr seqname	source	feature	start	end	score	strand	frame	[attributes]	[comments]
3_336_815_F3	solid	read	55409	55428	10.4	+	.	g=A3233312322232122211;i=1;p=1.000;q=23,12,18,17,10,24,19,14,27,9,23,9,16,20,11,7,8,4,4,14;u=0,0,0,1
3_142_1011_F3	solid	read	91290	91309	5.0	-	.	g=T0330222333132222222;i=1;p=1.000;q=4,4,14,4,4,4,4,21,4,4,4,4,25,4,4,4,5,21,4,4;u=0,0,0,1
3_341_424_F3	solid	read	102717	102736	10.6	-	.	g=T2203031313223113212;i=1;p=1.000;q=9,27,25,16,18,9,27,26,23,13,14,25,27,5,24,5,26,26,4,5;u=0,0,1
3_6_37_F3	solid	read	181053	181072	9.4	+	.	g=C3220221332111020310;i=1;p=1.000;q=9,5,13,9,10,22,6,12,21,7,13,4,21,16,23,6,20,20,13,6;u=0,0,0,1
3_34_202_F3	solid	read	284207	284226	6.9	+	.	g=G0301333332232122333;i=1;p=1.000;q=6,15,21,8,12,4,4,5,12,8,4,12,4,7,10,6,8,16,4,6;u=0,1
3_277_712_F3	solid	read	304136	304155	11.8	-	.	g=A2033101122223322133;i=1;p=1.000;q=26,11,14,27,4,17,4,26,26,23,17,25,26,27,21,23,5,20,26,23;u=0,1
3_394_71_F3	solid	read	308736	308755	10.8	+	.	g=T3203322323203312331;i=1;p=1.000;q=9,24,19,15,20,18,20,10,13,13,11,21,12,7,4,11,20,24,4,25;u=0,1
3_285_1497_F3	solid	read	404055	404074	8.4	-	.	g=T1221231003202232221;i=1;p=1.000;q=8,10,6,25,16,14,23,27,8,14,21,19,5,4,4,6,22,12,4,6;u=0,0,0,1
3_228_178_F3	solid	read	453227	453246	9.5	-	.	g=G1130333332331110323;i=1;p=1.000;q=4,19,25,18,18,5,19,6,8,24,4,26,21,11,15,4,26,13,13,15;u=0,0,0,1
3_406_794_F3	solid	read	504835	504854	8.3	-	.	g=T3033331301320201111;i=1;p=1.000;q=27,4,13,4,21,11,7,11,5,26,10,8,9,4,6,18,9,26,17,6;u=0,0,0,1
3_303_251_F3	solid	read	561501	561520	5.3	+	.	g=C0011111112222112221;i=1;p=1.000;q=9,8,4,4,10,4,4,4,6,14,4,4,4,4,16,4,4,4,4,23;u=0,0,1
3_152_112_F3	solid	read	624012	624031	7.7	-	.	g=G0301122312213122221;i=1;p=1.000;q=22,14,7,13,18,5,11,4,15,6,6,11,4,8,15,5,10,4,6,24;u=0,0,0,1
3_112_1154_F3	solid	read	630582	630601	11.3	-	.	g=T1333312011131131011;i=1;p=1.000;q=27,27,4,5,17,24,20,19,7,4,25,17,18,15,22,23,17,25,16,26;u=0,0,1
3_196_392_F3	solid	read	661664	661683	19.7	-	.	g=T3321013301122133323;i=1;p=1.000;q=27,25,13,26,21,25,23,27,27,27,27,11,16,27,27,19,26,27,26,27;u=1
3_192_1248_F3	solid	read	672037	672056	4.5	-	.	g=A0333232333121222222;i=1;p=1.000;q=4,7,4,4,4,4,4,4,6,4,4,4,4,4,7,7,4,4,6,4;u=0,0,0,1
3_63_479_F3	solid	read	742582	742601	7.9	-	.	g=A0133333333233232332;i=1;p=1.000;q=4,9,6,11,20,12,11,9,13,20,18,4,4,14,9,15,4,6,21,4;u=0,0,0,1
3_30_710_F3	solid	read	816069	816088	9.2	-	.	g=T3311001223313333313;i=1;p=1.000;q=22,27,18,25,25,7,26,25,14,23,6,25,5,11,7,4,15,7,4,6;u=0,0,0,1
3_284_77_F3	solid	read	864876	864895	7.4	+	.	g=T2003133033233112331;i=1;p=1.000;q=13,19,4,11,22,24,6,16,4,6,13,4,12,18,4,6,7,11,4,5;u=0,0,0,1
3_411_1040_F3	solid	read	876023	876042	10.9	-	.	g=T2121301233200033221;i=1;p=1.000;q=9,9,5,12,11,8,4,16,27,27,18,21,24,9,18,24,21,9,23,17;u=0,0,0,1
3_188_171_F3	solid	read	884683	884702	5.8	-	.	g=A1322330132213322231;i=1;p=1.000;q=4,8,4,5,7,6,5,4,11,6,6,11,4,8,4,8,4,6,4,15;u=0,0,0,1
3_63_787_F3	solid	read	1022149	1022168	7.5	+	.	g=C3131132013020123031;i=1;p=1.000;q=12,13,26,14,9,9,13,14,4,7,8,5,11,4,17,4,4,6,4,21;u=0,1
3_391_2015_F3	solid	read	1074989	1075008	18.5	-	.	g=A2323101222321232322;i=1;p=1.000;q=27,25,18,20,27,27,24,23,27,23,27,25,19,26,12,26,9,21,27,21;u=1
3_8_425_F3	solid	read	1119124	1119143	6.7	-	.	g=T0321201132230303323;i=1;p=1.000;q=6,5,8,6,4,4,23,9,12,10,15,4,13,13,8,4,4,5,5,12;u=0,0,1
3_53_745_F3	solid	read	1130179	1130198	7.6	-	.	g=C0213313233333113321;i=1;p=1.000;q=27,6,9,22,18,9,8,15,6,8,14,5,8,6,16,4,5,4,4,14;u=0,0,0,1
3_123_576_F3	solid	read	1219122	1219141	8.7	+	.	g=A3333133323333323323;i=1;p=1.000;q=18,22,5,11,16,16,8,14,8,5,19,8,9,10,7,11,6,11,9,4;u=0,0,1
3_81_12_F3	solid	read	1236732	1236751	8.6	+	.	g=G2210332302233112321;i=1;p=1.000;q=7,16,17,9,7,9,9,16,9,4,10,21,17,8,4,6,9,16,6,12;u=0,0,0,1
3_96_1862_F3	solid	read	1264409	1264428	6.9	-	.	g=G0301032323231222021;i=1;p=1.000;q=26,23,11,20,15,8,6,4,6,6,9,7,6,4,8,6,4,5,6,5;u=0,0,0,1
3_40_136_F3	solid	read	1266177	1266196	7.4	-	.	g=T2332222332203312221;i=1;p=1.000;q=9,23,6,19,13,9,4,8,17,9,4,4,13,9,8,5,4,6,10,8;u=0,0,1
3_124_1781_F3	solid	read	1385416	1385435	10.3	+	.	g=A1322302333332222132;i=1;p=1.000;q=13,17,8,6,5,9,24,4,7,9,18,27,18,16,16,23,18,18,11,23;u=0,0,1
3_134_1165_F3	solid	read	1393169	1393188	9.0	-	.	g=T3301123202321131311;i=1;p=1.000;q=4,27,18,7,27,4,27,26,4,20,4,27,26,9,27,4,27,14,10,27;u=1
3_224_587_F3	solid	read	1490044	1490063	6.1	+	.	g=G2032313231111233321;i=1;p=1.000;q=4,4,6,6,13,24,4,4,5,15,6,7,9,14,4,4,4,25,5,5;u=0,0,0,1
3_25_747_F3	solid	read	1513598	1513617	9.5	+	.	g=T1223213101133121231;i=1;p=1.000;q=26,27,8,27,27,27,26,27,26,19,8,14,4,17,11,5,7,4,7,6;u=0,0,1
3_143_14_F3	solid	read	1528236	1528255	9.7	+	.	g=T3233113323230202011;i=1;p=1.000;q=13,23,17,19,23,16,24,25,14,15,9,6,4,11,4,9,12,4,16,10;u=0,0,0,1
3_164_1025_F3	solid	read	1570107	1570126	7.9	-	.	g=T3220332323303320231;i=1;p=1.000;q=7,10,20,8,4,24,4,4,21,6,26,22,9,6,11,9,6,4,17,14;u=0,0,0,1
3_137_552_F3	solid	read	1630276	1630295	9.1	-	.	g=G3030333223233102131;i=1;p=1.000;q=6,28,9,4,6,26,27,6,10,9,27,21,6,16,9,25,6,7,23,12;u=0,0,0,1
3_125_1810_F3	solid	read	1634104	1634123	10.5	+	.	g=G1232220322032311332;i=1;p=1.000;q=27,8,26,26,10,6,26,12,27,27,26,4,27,27,23,8,8,4,27,12;u=0,0,0,1
3_314_1310_F3	solid	read	1639981	1640000	9.2	+	.	g=A2221332230322203033;i=1;p=1.000;q=19,12,6,27,11,27,6,11,5,6,9,13,27,27,8,18,5,22,4,27;u=0,0,0,1
3_384_591_F3	solid	read	1654341	1654360	6.8	+	.	g=A3323221133121102313;i=1;p=1.000;q=19,8,7,7,15,4,20,7,4,6,14,7,19,6,8,4,5,9,4,4;u=0,0,0,1
3_145_739_F3	solid	read	1791040	1791059	11.9	-	.	g=A0221223333323131212;i=1;p=1.000;q=20,27,23,13,27,14,27,28,27,25,12,24,8,16,8,4,8,21,9,11;u=0,0,0,1
3_326_2020_F3	solid	read	1830564	1830583	9.3	+	.	g=A3321322331103233322;i=1;p=1.000;q=14,4,25,16,10,12,16,5,14,10,25,5,25,5,9,18,13,26,4,26;u=0,0,0,1
3_233_1265_F3	solid	read	1857564	1857583	8.9	+	.	g=T3112113020130223311;i=1;p=1.000;q=7,27,25,26,27,14,26,27,27,27,4,6,5,10,17,4,5,7,6,12;u=0,0,1
3_235_100_F3	solid	read	1912460	1912479	9.6	-	.	g=G2233020000132311231;i=1;p=1.000;q=23,24,25,16,17,6,21,25,9,4,6,11,8,19,6,6,19,14,13,6;u=0,0,0,1
3_111_107_F3	solid	read	1944496	1944515	7.6	-	.	g=C3023223333211322231;i=1;p=1.000;q=15,5,6,14,5,13,4,12,11,4,9,9,11,12,4,11,11,13,6,6;u=0,0,0,1
3_457_1514_F3	solid	read	1956598	1956617	9.9	-	.	g=T0013331013332110221;i=1;p=1.000;q=18,24,10,24,23,25,22,11,20,10,15,11,4,5,27,4,9,13,5,27;u=0,1
3_183_74_F3	solid	read	1992040	1992059	9.8	+	.	g=C3332233131131222322;i=1;p=1.000;q=27,27,25,23,25,8,11,11,7,11,4,12,14,10,15,7,14,4,9,12;u=0,0,1
3_357_1303_F3	solid	read	2037917	2037936	10.9	-	.	g=T3331331323320311331;i=1;p=1.000;q=7,27,5,19,26,8,27,12,14,27,8,27,23,9,19,4,26,20,9,27;u=0,0,0,1
3_153_186_F3	solid	read	2083441	2083460	6.7	+	.	g=T3112233331133323322;i=1;p=1.000;q=7,14,19,7,12,6,11,4,11,8,4,6,6,4,11,4,6,4,4,18;u=0,1
3_65_1741_F3	solid	read	2107441	2107460	8.4	+	.	g=T3333332330233132123;i=1;p=1.000;q=4,4,6,25,9,4,26,16,21,9,18,15,27,27,4,21,9,7,9,6;u=0,0,0,1
3_98_323_F3	solid	read	2118821	2118840	7.5	+	.	g=A3222212322131112031;i=1;p=1.000;q=13,14,8,10,8,14,4,13,10,7,15,4,6,4,4,12,6,11,6,8;u=0,0,1
3_48_258_F3	solid	read	2153882	2153901	9.4	-	.	g=G0330113313201122321;i=1;p=1.000;q=22,15,20,4,16,17,14,24,4,5,4,22,19,8,10,9,13,22,8,15;u=0,0,0,1
3_140_1125_F3	solid	read	2182909	2182928	7.9	+	.	g=T3231331302232001131;i=1;p=1.000;q=10,4,12,6,4,12,13,6,18,5,8,11,4,26,6,25,5,18,11,12;u=0,0,0,1
3_359_118_F3	solid	read	2188393	2188412	8.4	+	.	g=A0301311133331131322;i=1;p=1.000;q=11,5,7,13,20,6,6,25,8,18,9,15,27,9,6,7,15,17,4,4;u=0,0,0,1
3_203_483_F3	solid	read	2272874	2272893	9.1	-	.	g=C3031223110333133311;i=1;p=1.000;q=23,21,25,27,10,5,22,15,17,18,5,18,17,5,19,4,4,13,4,22;u=0,0,0,1
3_66_301_F3	solid	read	2286038	2286057	6.6	-	.	g=C1113113330132222311;i=1;p=1.000;q=10,4,6,4,8,13,9,4,10,9,4,6,13,9,5,6,11,6,4,9;u=0,0,0,1
3_78_130_F3	solid	read	2291021	2291040	7.6	+	.	g=G3233131332212222321;i=1;p=1.000;q=13,16,6,12,17,11,10,4,12,8,13,4,8,6,4,4,12,10,4,11;u=0,0,0,1
3_141_110_F3	solid	read	2291354	2291373	9.3	+	.	g=T1312203322212123321;i=1;p=1.000;q=9,21,24,11,16,4,23,27,16,16,8,22,6,10,16,4,9,4,7,25;u=0,0,1
3_51_1383_F3	solid	read	2374918	2374937	8.8	+	.	g=T3311203033322222231;i=1;p=1.000;q=24,26,6,27,27,23,27,4,21,27,4,27,6,9,24,4,23,4,4,27;u=0,0,1
3_231_366_F3	solid	read	2392091	2392110	10.0	-	.	g=T2022333223101331322;i=1;p=1.000;q=18,12,9,9,13,8,7,22,7,7,4,26,12,17,9,20,24,8,18,14;u=0,0,0,1
3_214_1802_F3	solid	read	2394604	2394623	8.8	-	.	g=T1232111001220211133;i=1;p=1.000;q=17,18,14,6,19,4,21,4,6,12,11,4,26,20,9,18,7,16,5,18;u=0,0,0,1
3_67_1434_F3	solid	read	2454508	2454527	15.2	-	.	g=T3121311232222231203;i=1;p=1.000;q=9,27,27,18,16,14,25,27,26,21,19,27,27,27,15,5,24,27,24,24;u=0,0,1
3_124_1647_F3	solid	read	2493617	2493636	7.5	+	.	g=A0211320203220231332;i=1;p=1.000;q=9,12,12,9,6,14,12,7,4,4,12,9,4,9,16,4,4,9,9,16;u=0,0,0,1
3_39_328_F3	solid	read	2500759	2500778	7.8	+	.	g=T1332333033231132333;i=1;p=1.000;q=24,27,26,26,25,21,7,8,4,5,20,4,11,6,8,4,6,4,11,7;u=0,0,1
3_378_322_F3	solid	read	2541624	2541643	8.9	+	.	g=T2333331001023011220;i=1;p=1.000;q=14,6,13,25,27,4,24,22,14,19,9,23,15,6,8,4,22,4,4,20;u=0,0,0,1
3_216_848_F3	solid	read	2550573	2550592	11.5	-	.	g=G2320322020031220322;i=1;p=1.000;q=21,24,8,21,20,25,18,6,24,14,21,9,7,18,8,18,7,9,19,12;u=0,0,0,1
3_221_516_F3	solid	read	2607559	2607578	11.1	-	.	g=T2132333313222333332;i=1;p=1.000;q=9,19,27,26,24,26,26,25,25,26,21,4,6,10,21,6,20,13,5,24;u=0,0,0,1
3_56_45_F3	solid	read	2662103	2662122	5.5	+	.	g=G3021122332232122321;i=1;p=1.000;q=4,4,4,6,4,6,4,5,18,9,4,16,10,4,4,4,12,4,6,6;u=0,0,0,1
3_127_210_F3	solid	read	2798906	2798925	10.2	+	.	g=G2331321333232203222;i=1;p=1.000;q=11,25,9,4,23,16,26,14,7,22,9,25,9,8,21,8,15,17,4,26;u=0,0,1
3_417_422_F3	solid	read	2812322	2812341	8.8	-	.	g=T3321222333313333132;i=1;p=1.000;q=9,26,7,19,7,13,23,4,25,4,6,19,4,16,15,15,23,4,19,13;u=0,0,0,1
3_42_1403_F3	solid	read	2830264	2830283	9.6	-	.	g=T3212330132120221212;i=1;p=1.000;q=7,4,25,18,6,17,12,12,17,14,8,26,13,15,10,4,21,5,12,22;u=0,1
3_457_42_F3	solid	read	2874245	2874264	7.6	-	.	g=G0301123332223122221;i=1;p=1.000;q=18,10,14,9,19,4,10,8,11,10,6,8,5,8,11,4,13,6,4,6;u=0,0,1
3_361_728_F3	solid	read	2893879	2893898	14.6	+	.	g=C3213223312310132221;i=1;p=1.000;q=14,18,7,7,17,19,23,24,17,26,12,15,21,23,21,19,17,20,22,24;u=0,0,0,1
3_77_718_F3	solid	read	2913092	2913111	9.4	+	.	g=T3021331333313131231;i=1;p=1.000;q=15,26,7,24,20,18,5,6,17,18,6,11,4,13,19,15,7,4,22,25;u=0,0,0,1
3_116_154_F3	solid	read	2917672	2917691	9.8	-	.	g=A0323231223233132311;i=1;p=1.000;q=20,9,19,18,10,18,8,16,25,6,18,6,12,24,6,7,5,15,7,17;u=0,0,0,1
3_239_1415_F3	solid	read	2923256	2923275	19.2	+	.	g=T3233113121300032200;i=1;p=1.000;q=25,27,27,26,27,24,27,27,25,27,22,27,21,26,22,19,26,9,14,21;u=1
3_142_1468_F3	solid	read	2930117	2930136	10.5	-	.	g=A3233323333303103330;i=1;p=1.000;q=9,20,6,26,16,18,8,13,20,25,25,18,6,12,11,18,4,16,16,6;u=0,0,1
3_394_295_F3	solid	read	2930118	2930137	8.1	-	.	g=T3023333333333311331;i=1;p=1.000;q=4,14,6,12,7,22,10,4,13,24,18,12,12,4,6,9,9,9,14,4;u=0,0,0,1
3_222_1773_F3	solid	read	2934040	2934059	11.6	+	.	g=T1303031311123232302;i=1;p=1.000;q=11,10,24,15,28,6,19,5,13,27,8,26,8,22,25,27,26,27,8,13;u=0,0,0,1
3_276_1344_F3	solid	read	2969950	2969969	13.2	-	.	g=G3211212131233322233;i=1;p=1.000;q=27,27,12,16,11,23,27,8,23,12,27,22,20,12,15,25,8,27,16,6;u=0,1
3_155_1814_F3	solid	read	3107393	3107412	13.6	+	.	g=A2332222213113120221;i=1;p=1.000;q=27,26,20,25,26,27,12,27,26,18,26,4,27,10,23,26,6,23,26,26;u=0,0,0,1
3_373_2014_F3	solid	read	3143956	3143975	12.0	-	.	g=T3013322223222221211;i=1;p=1.000;q=16,8,17,21,10,10,18,18,18,13,4,23,16,24,8,19,14,15,23,11;u=0,1
3_81_1637_F3	solid	read	3413619	3413638	9.1	+	.	g=G2313032322122302111;i=1;p=1.000;q=9,4,7,19,27,6,11,5,12,15,20,27,8,27,6,16,6,27,21,6;u=0,0,1
3_291_969_F3	solid	read	3438323	3438342	17.4	+	.	g=T0021120212032121313;i=1;p=1.000;q=24,27,6,27,27,27,27,13,27,27,25,27,26,27,27,20,23,26,27,20;u=1
3_179_1617_F3	solid	read	3475164	3475183	8.0	+	.	g=A2100132222332123123;i=1;p=1.000;q=21,25,11,22,4,19,7,21,20,4,5,24,25,16,4,4,11,19,4,4;u=0,0,0,1
3_446_861_F3	solid	read	3476173	3476192	11.6	-	.	g=G1213302212022132321;i=1;p=1.000;q=27,27,27,27,26,25,12,27,24,18,24,6,27,26,20,9,6,6,4,23;u=0,0,1
3_397_317_F3	solid	read	3545152	3545171	11.1	+	.	g=T3110031332233111131;i=1;p=1.000;q=22,27,9,9,26,5,22,20,9,10,16,22,24,6,23,25,22,4,17,18;u=0,0,0,1
3_323_713_F3	solid	read	3575287	3575306	16.2	-	.	g=A0322222200213223302;i=1;p=1.000;q=27,25,21,27,26,26,24,26,27,18,27,26,26,27,22,22,6,26,25,8;u=0,1
3_294_1906_F3	solid	read	3727542	3727561	8.4	-	.	g=A3030310223202311021;i=1;p=1.000;q=14,7,5,4,7,18,4,6,13,6,12,12,10,11,15,14,16,7,9,12;u=0,0,0,1
3_443_223_F3	solid	read	3730805	3730824	17.1	-	.	g=T1113320033330133111;i=1;p=1.000;q=28,27,18,27,27,27,20,26,27,14,25,16,19,19,8,23,16,21,16,15;u=0,0,1
3_94_809_F3	solid	read	3841898	3841917	21.8	-	.	g=A2032223110001131310;i=1;p=1.000;q=27,27,27,27,26,27,25,24,27,27,27,25,27,27,27,12,23,16,27,27;u=0,0,0,1
3_245_387_F3	solid	read	3878549	3878568	24.4	-	.	g=A0222211220333132122;i=1;p=1.000;q=27,27,26,27,26,27,27,25,27,25,26,27,18,21,26,25,26,23,24,24;u=1
3_190_1089_F3	solid	read	3900038	3900057	13.7	-	.	g=T1111110323122301202;i=1;p=1.000;q=27,11,27,11,8,9,27,9,9,26,25,27,11,27,23,14,24,20,22,26;u=0,0,1
3_442_1501_F3	solid	read	3912610	3912629	8.5	+	.	g=A0012333103302132301;i=1;p=1.000;q=11,11,15,19,15,6,12,10,4,11,21,5,9,16,7,14,4,4,8,19;u=0,0,1
3_342_678_F3	solid	read	4044575	4044594	4.0	+	.	g=A3333112332213322323;i=1;p=1.000;q=4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4;u=0,0,0,1
3_56_1294_F3	solid	read	4058789	4058808	12.7	+	.	g=G3323331232322213322;i=1;p=1.000;q=26,17,18,27,23,8,8,24,27,27,9,27,25,14,26,4,27,9,24,23;u=0,0,0,1
3_69_1575_F3	solid	read	4070467	4070486	9.9	+	.	g=A2222011012222112121;i=1;p=1.000;q=16,25,14,9,9,9,21,9,4,24,6,21,13,6,27,10,19,8,6,27;u=0,0,0,1
3_198_476_F3	solid	read	4080622	4080641	8.9	+	.	g=C2010231122212011133;i=1;p=1.000;q=16,8,8,16,12,17,4,16,12,15,10,4,9,6,4,25,9,9,23,11;u=0,1
3_24_715_F3	solid	read	4136503	4136522	4.0	-	.	g=G1313332132232313233;i=1;p=1.000;q=4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4;u=0,0,0,1
3_151_283_F3	solid	read	4148264	4148283	9.7	+	.	g=T3230210232022111220;i=1;p=1.000;q=9,14,6,25,25,19,6,4,16,11,12,20,10,13,26,19,6,4,19,14;u=0,0,1
3_164_774_F3	solid	read	4156157	4156176	9.6	+	.	g=G2311112210110223313;i=1;p=1.000;q=8,24,19,7,6,16,12,9,4,8,26,14,26,24,7,18,6,16,14,7;u=0,0,0,1
3_275_1212_F3	solid	read	4171385	4171404	8.3	+	.	g=G0223122231333302232;i=1;p=1.000;q=13,8,5,4,10,7,12,25,4,25,6,15,6,27,6,11,12,7,14,10;u=0,0,0,1
3_148_289_F3	solid	read	4177672	4177691	8.0	-	.	g=T1203101332223323323;i=1;p=1.000;q=9,21,11,6,5,7,25,24,26,24,8,9,7,12,7,4,11,9,4,4;u=0,0,0,1
3_437_1000_F3	solid	read	4179623	4179642	12.3	+	.	g=A0112222212231131001;i=1;p=1.000;q=26,27,26,27,4,27,17,6,22,13,27,24,6,27,21,27,22,15,24,9;u=0,0,1
3_318_2011_F3	solid	read	4218181	4218200	12.9	-	.	g=T2133330223033303323;i=1;p=1.000;q=25,27,27,5,5,16,27,16,27,15,18,25,26,11,27,19,16,24,9,15;u=0,0,0,1
3_14_11_F3	solid	read	4222697	4222716	7.8	-	.	g=T2323310222232322122;i=1;p=1.000;q=6,23,16,25,25,9,7,4,12,4,14,6,10,7,6,9,18,4,10,4;u=0,0,0,1
3_402_391_F3	solid	read	4274545	4274564	6.2	-	.	g=C3303323321111111111;i=1;p=1.000;q=10,19,15,15,7,8,13,4,7,4,5,16,4,4,5,4,9,4,4,4;u=0,0,0,1
3_293_504_F3	solid	read	4339235	4339254	9.5	+	.	g=C2133223303331120213;i=1;p=1.000;q=6,4,5,26,13,7,17,6,24,10,27,24,5,9,21,9,23,24,20,14;u=0,0,0,1
3_360_914_F3	solid	read	4407004	4407023	10.7	+	.	g=T3012102130232022001;i=1;p=1.000;q=23,24,19,17,24,6,26,17,25,15,7,24,14,11,26,9,22,4,8,5;u=0,0,0,1
3_118_1532_F3	solid	read	4431702	4431721	10.2	+	.	g=C3233220201223200322;i=1;p=1.000;q=20,9,17,22,17,23,13,4,9,5,16,11,10,6,17,7,9,22,27,27;u=0,0,1
3_358_133_F3	solid	read	4460191	4460210	9.1	+	.	g=T0221223112322112233;i=1;p=1.000;q=6,23,12,22,7,6,7,4,13,5,9,23,12,9,24,8,14,7,20,26;u=0,0,0,1
3_397_195_F3	solid	read	4499390	4499409	6.9	-	.	g=T3302332313332212121;i=1;p=1.000;q=23,14,15,5,9,8,6,4,4,13,4,16,13,16,4,7,4,12,4,5;u=0,0,0,1
3_158_642_F3	solid	read	4533144	4533163	7.1	-	.	g=A1332103332323233212;i=1;p=1.000;q=8,20,9,22,8,14,4,16,17,4,8,13,7,8,4,12,5,4,4,4;u=0,0,0,1
3_300_1439_F3	solid	read	4580452	4580471	12.3	-	.	g=A0331111211302100201;i=1;p=1.000;q=5,17,21,14,4,16,11,27,21,9,17,17,27,23,12,21,16,27,25,25;u=0,0,0,1
# Elapsed time 0.846 secs
</pre></div>
</div>
</div>
<div class="section" id="id16">
<h3>Import<a class="headerlink" href="#id16" title="Permalink to this headline">Â¶</a></h3>
<p>By default, the <code class="docutils literal"><span class="pre">id_spec=&quot;some_string&quot;</span></code> parameter tells <a class="reference internal" href="api.html#module-gffutils" title="gffutils"><code class="xref py py-mod docutils literal"><span class="pre">gffutils</span></code></a> to
create a primary key based on the value of the &#8220;some_string&#8221; attribute.  For
this example, however, we would like to make the first field the primary key.
Luckily, <a class="reference internal" href="api.html#module-gffutils" title="gffutils"><code class="xref py py-mod docutils literal"><span class="pre">gffutils</span></code></a> will accept the name of a GFF field, surrounded by
&#8221;:&#8221;.  The names of the fields are in <code class="docutils literal"><span class="pre">gffutils.constants._gffkeys</span></code>:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">gffutils</span><span class="o">.</span><span class="n">constants</span><span class="o">.</span><span class="n">_gffkeys</span>
<span class="go">[&#39;seqid&#39;, &#39;source&#39;, &#39;featuretype&#39;, &#39;start&#39;, &#39;end&#39;, &#39;score&#39;, &#39;strand&#39;, &#39;frame&#39;, &#39;attributes&#39;]</span>
</pre></div>
</div>
<p>We want to use the first field, <code class="docutils literal"><span class="pre">seqid</span></code>, as the primary key for the
database, so we use <code class="docutils literal"><span class="pre">id_spec=&quot;:seqid:&quot;</span></code>.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fn</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span><span class="s">&#39;F3-unique-3.v2.gff&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">create_db</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="s">&#39;:memory:&#39;</span><span class="p">,</span> <span class="n">id_spec</span><span class="o">=</span><span class="p">[</span><span class="s">&#39;:seqid:&#39;</span><span class="p">])</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">db</span><span class="o">.</span><span class="n">directives</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">&#39;solid-gff-version 0.2&#39;</span>
</pre></div>
</div>
<p>Now we can access the features by their sequence name:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">db</span><span class="p">[</span><span class="s">&#39;3_8_425_F3&#39;</span><span class="p">]</span>
<span class="go">&lt;Feature read (3_8_425_F3:1119124-1119143[-]) at 0x...&gt;</span>
</pre></div>
</div>
</div>
</div>
<div class="html-toggle section" id="glimmer-nokeyval-gff3">
<span id="id17"></span><h2>glimmer_nokeyval.gff3<a class="headerlink" href="#glimmer-nokeyval-gff3" title="Permalink to this headline">Â¶</a></h2>
<p>This is a problematic file in which the attributes are not all in format
&#8220;key=val&#8221;.  This example demonstrates how <a class="reference internal" href="api.html#module-gffutils" title="gffutils"><code class="xref py py-mod docutils literal"><span class="pre">gffutils</span></code></a> handles cases like
this: by implicitly treating them as keys and setting the value to an empty
list.</p>
<p>Like the the <a class="reference internal" href="#ensembl-gtf-txt"><span>ensembl_gtf.txt</span></a> file, gene and transcript names are not
different.  Based on IDs alone, it&#8217;s not clear whether the CDS&#8217;s parent is the
gene or the mRNA.</p>
<p>However, the way to fix this is different from <code class="docutils literal"><span class="pre">ensembl_gtf.txt</span></code>.  Recall in
that case, we constructed a function to modify transcript IDs, and when
<a class="reference internal" href="api.html#module-gffutils" title="gffutils"><code class="xref py py-mod docutils literal"><span class="pre">gffutils</span></code></a> inferred the new transcripts from the component exons, it used
that new ID.</p>
<p>Here, we need to do several things in the transform function:  if the feature
type contains &#8220;RNA&#8221;, then set the ID to <code class="docutils literal"><span class="pre">ID</span> <span class="pre">+</span> <span class="pre">&quot;_transript&quot;</span></code> (this will handle
things like ncRNA, tRNA, etc and is more general than just &#8220;mRNA&#8221;).  Next, we
need to edit the Parent attribute of CDS, exon, etc.  How this is handled will
depend completely on the file you are using.</p>
<div class="html-toggle section" id="id18">
<h3>File contents<a class="headerlink" href="#id18" title="Permalink to this headline">Â¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>##gff-version 3
##sequence-region scaffold4215_3 1 6526
scaffold4215_3	glimmer	gene	3	62	.	-	.	ID=GL0000006;Name=GL0000006;Lack 3&#39;-end;
scaffold4215_3	glimmer	mRNA	3	62	.	-	.	ID=GL0000006;Name=GL0000006;Parent=GL0000006;Lack 3&#39;-end;
scaffold4215_3	glimmer	CDS	3	62	2.84	-	0	Parent=GL0000006;Lack 3&#39;-end;
scaffold4215_3	glimmer	gene	124	1983	.	-	.	ID=GL0000007;Name=GL0000007;Complete;
</pre></div>
</div>
</div>
<div class="section" id="id19">
<h3>Import<a class="headerlink" href="#id19" title="Permalink to this headline">Â¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">transform</span><span class="p">(</span><span class="n">f</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">f</span><span class="o">.</span><span class="n">featuretype</span> <span class="o">==</span> <span class="s">&#39;gene&#39;</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">return</span> <span class="n">f</span>
<span class="gp">... </span>    <span class="k">elif</span> <span class="s">&quot;RNA&quot;</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">featuretype</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">f</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;ID&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+=</span> <span class="s">&#39;_transcript&#39;</span>
<span class="gp">... </span>    <span class="k">else</span><span class="p">:</span>
<span class="gp">... </span>        <span class="c"># assume that anything else is a child of a transcript, so we need</span>
<span class="gp">... </span>        <span class="c"># to edit the &quot;Parent&quot; attribute</span>
<span class="gp">... </span>        <span class="k">if</span> <span class="s">&#39;Parent&#39;</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">attributes</span><span class="p">:</span>
<span class="gp">... </span>            <span class="n">f</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;Parent&#39;</span><span class="p">]</span> \
<span class="gp">... </span>               <span class="o">=</span> <span class="p">[</span><span class="n">i</span> <span class="o">+</span> <span class="s">&#39;_transcript&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">f</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;Parent&#39;</span><span class="p">]]</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">f</span>
<span class="gp">...</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fn</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span><span class="s">&#39;glimmer_nokeyval.gff3&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">create_db</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="s">&quot;:memory:&quot;</span><span class="p">,</span> <span class="n">id_spec</span><span class="o">=</span><span class="s">&quot;ID&quot;</span><span class="p">,</span> <span class="n">transform</span><span class="o">=</span><span class="n">transform</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="id20">
<h3>Access<a class="headerlink" href="#id20" title="Permalink to this headline">Â¶</a></h3>
<p>This shows that keys with missing values are assigned an empty list.  It&#8217;s up
to the calling code to decide how to handle this (say, by checking for the
presence of the key &#8220;Complete&#8221;.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">db</span><span class="p">[</span><span class="s">&#39;GL0000007&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">attributes</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="s">&#39;=&#39;</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span>
<span class="go">Complete = []</span>
<span class="go">ID = [&#39;GL0000007&#39;]</span>
<span class="go">Name = [&#39;GL0000007&#39;]</span>
</pre></div>
</div>
<p>This illustrates that the CDS finds the proper transcript parent.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">db</span><span class="o">.</span><span class="n">parents</span><span class="p">(</span><span class="n">db</span><span class="p">[</span><span class="s">&#39;CDS_1&#39;</span><span class="p">],</span> <span class="n">level</span><span class="o">=</span><span class="mi">1</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
<span class="go">GL0000006_transcript</span>
</pre></div>
</div>
</div>
</div>
<div class="html-toggle section" id="hybrid1-gff3">
<span id="id21"></span><h2>hybrid1.gff3<a class="headerlink" href="#hybrid1-gff3" title="Permalink to this headline">Â¶</a></h2>
<p>This file contains a FASTA sequence at the end.  Currently, <a class="reference internal" href="api.html#module-gffutils" title="gffutils"><code class="xref py py-mod docutils literal"><span class="pre">gffutils</span></code></a>
assumes that there are no further annotations and stops parsing the file at
this point (specifically, when it sees the string &#8220;##FASTA&#8221; at the beginning of
a line).</p>
<p>Another issue with this file is that the &#8220;Note&#8221; attributes field contains
url-encoded characters. These are kept verbatim.</p>
<div class="section" id="id22">
<h3>File contents<a class="headerlink" href="#id22" title="Permalink to this headline">Â¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>##gff-version 3
##sequence-region foo 1 100
##feature-ontology bar
##attribute-ontology baz
##source-ontology boo
##sequence-region chr17 62467934 62469545 
chr17	UCSC	mRNA	62467934	62469545	.	-	.	ID=A00469;Dbxref=AFFX-U133:205840_x_at,Locuslink:2688,Genbank-mRNA:A00469,Swissprot:P01241,PFAM:PF00103,AFFX-U95:1332_f_at,Swissprot:SOMA_HUMAN;Note=growth%20hormone%201;Alias=GH1
chr17	UCSC	CDS	62468039	62468236	.	-	1	Parent=A00469
chr17	UCSC	CDS	62468490	62468654	.	-	2	Parent=A00469
chr17	UCSC	CDS	62468747	62468866	.	-	1	Parent=A00469
chr17	UCSC	CDS	62469076	62469236	.	-	1	Parent=A00469
chr17	UCSC	CDS	62469497	62469506	.	-	0	Parent=A00469
###
##FASTA
&gt;chr17
GATTACA
GATTACA
</pre></div>
</div>
</div>
<div class="section" id="id23">
<h3>Import<a class="headerlink" href="#id23" title="Permalink to this headline">Â¶</a></h3>
<p>Straightforward:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fn</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span><span class="s">&#39;hybrid1.gff3&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">create_db</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="s">&quot;:memory:&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="id24">
<h3>Access<a class="headerlink" href="#id24" title="Permalink to this headline">Â¶</a></h3>
<p>Print out the attributes:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">db</span><span class="p">[</span><span class="s">&#39;A00469&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">attributes</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="n">k</span><span class="p">,</span> <span class="s">&#39;=&#39;</span><span class="p">,</span> <span class="n">v</span><span class="p">)</span>
<span class="go">Alias = [&#39;GH1&#39;]</span>
<span class="go">Dbxref = [&#39;AFFX-U133:205840_x_at&#39;, &#39;Locuslink:2688&#39;, &#39;Genbank-mRNA:A00469&#39;, &#39;Swissprot:P01241&#39;, &#39;PFAM:PF00103&#39;, &#39;AFFX-U95:1332_f_at&#39;, &#39;Swissprot:SOMA_HUMAN&#39;]</span>
<span class="go">ID = [&#39;A00469&#39;]</span>
<span class="go">Note = [&#39;growth%20hormone%201&#39;]</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="jgi-gff2-txt">
<span id="id25"></span><h2>jgi_gff2.txt<a class="headerlink" href="#jgi-gff2-txt" title="Permalink to this headline">Â¶</a></h2>
<p>This file is GFF2 format &#8211; the attribute format is like GTF, but without the
required &#8220;gene_id&#8221; and &#8220;transcript_id&#8221; fields that are used to infer genes and
transcripts. To get around this, we can supply the <code class="docutils literal"><span class="pre">gtf_transcript_key</span></code> kwarg,
which will override the default <code class="docutils literal"><span class="pre">&quot;transcript_id&quot;</span></code> (see <a class="reference internal" href="gtf.html#gtf"><span>GTF files</span></a> for more
background on this).</p>
<p>There is information in the file on which exon goes to which transcript, but no
such information about which CDS goes to which transcript.  There is, however,
information about the parent protein.</p>
<div class="section" id="id26">
<h3>File contents<a class="headerlink" href="#id26" title="Permalink to this headline">Â¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>chr_1	JGI	exon	37061	37174	.	-	.	name &quot;fgenesh1_pg.C_chr_1000007&quot;; transcriptId 873
chr_1	JGI	CDS	37061	37174	.	-	0	name &quot;fgenesh1_pg.C_chr_1000007&quot;; proteinId 873; exonNumber 3
chr_1	JGI	exon	37315	37620	.	-	.	name &quot;fgenesh1_pg.C_chr_1000007&quot;; transcriptId 873
chr_1	JGI	CDS	37315	37620	.	-	0	name &quot;fgenesh1_pg.C_chr_1000007&quot;; proteinId 873; exonNumber 2
chr_1	JGI	exon	37752	38216	.	-	.	name &quot;fgenesh1_pg.C_chr_1000007&quot;; transcriptId 873
chr_1	JGI	CDS	37752	38216	.	-	0	name &quot;fgenesh1_pg.C_chr_1000007&quot;; proteinId 873; exonNumber 1
</pre></div>
</div>
</div>
<div class="section" id="id27">
<h3>Import<a class="headerlink" href="#id27" title="Permalink to this headline">Â¶</a></h3>
<p>Inspecting the file, we see that the &#8220;name&#8221; attribute is repeated on each line.
We can use that as the gene ID for inferring the gene extents, like &#8220;gene_id&#8221;
in a normal GTF file.  Similarly, the &#8220;transcriptId&#8221; attribute here can be used
like &#8220;transcript_id&#8221; in a normal GTF file for inferring transcript extents.</p>
<p>We also need to make a choice about how we&#8217;re going to use this database.  Do
we want to have the CDS features be considered children of transcripts?  In
that case, we&#8217;ll need the transform function, which creates a &#8220;transcriptId&#8221;
attribute out of an existing &#8220;proteinId&#8221; attribute:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">transform</span><span class="p">(</span><span class="n">d</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">try</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">d</span><span class="p">[</span><span class="s">&#39;transcriptId&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">d</span><span class="p">[</span><span class="s">&#39;proteinId&#39;</span><span class="p">]</span>
<span class="gp">... </span>    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">pass</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">d</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fn</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span><span class="s">&#39;jgi_gff2.txt&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">create_db</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="s">&quot;:memory:&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">id_spec</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;transcript&#39;</span><span class="p">:</span> <span class="s">&#39;transcriptId&#39;</span><span class="p">,</span> <span class="s">&#39;gene&#39;</span><span class="p">:</span> <span class="s">&#39;name&#39;</span><span class="p">},</span>
<span class="gp">... </span><span class="n">gtf_transcript_key</span><span class="o">=</span><span class="s">&#39;transcriptId&#39;</span><span class="p">,</span> <span class="n">gtf_gene_key</span><span class="o">=</span><span class="s">&#39;name&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">transform</span><span class="o">=</span><span class="n">transform</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="id28">
<h3>Access<a class="headerlink" href="#id28" title="Permalink to this headline">Â¶</a></h3>
<p>Since we used that transform function, the exons and CDSs are all children of
the &#8220;873&#8221; transcript:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="n">db</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="s">&#39;873&#39;</span><span class="p">,</span> <span class="n">level</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
<span class="go">[&lt;Feature CDS (chr_1:37061-37174[-]) at 0x...&gt;, &lt;Feature CDS (chr_1:37315-37620[-]) at 0x...&gt;, &lt;Feature CDS (chr_1:37752-38216[-]) at 0x...&gt;, &lt;Feature exon (chr_1:37061-37174[-]) at 0x...&gt;, &lt;Feature exon (chr_1:37315-37620[-]) at 0x...&gt;, &lt;Feature exon (chr_1:37752-38216[-]) at 0x...&gt;]</span>
</pre></div>
</div>
<p>Here we can see that all children of the gene are accounted for:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">db</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="s">&quot;fgenesh1_pg.C_chr_1000007&quot;</span><span class="p">,</span> <span class="n">order_by</span><span class="o">=</span><span class="s">&#39;featuretype&#39;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="s">&#39;{0.featuretype:&gt;10}: {0.id}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">f</span><span class="p">))</span>
<span class="go">       CDS: CDS_1</span>
<span class="go">       CDS: CDS_2</span>
<span class="go">       CDS: CDS_3</span>
<span class="go">      exon: exon_1</span>
<span class="go">      exon: exon_2</span>
<span class="go">      exon: exon_3</span>
<span class="go">transcript: 873</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="ncbi-gff3-txt">
<span id="id29"></span><h2>ncbi_gff3.txt<a class="headerlink" href="#ncbi-gff3-txt" title="Permalink to this headline">Â¶</a></h2>
<p>This file is problematic because all genes have the same ID attribute.  We can
get around this by using the &#8220;db_xref&#8221; field as the key for genes.</p>
<div class="section" id="id30">
<h3>File contents<a class="headerlink" href="#id30" title="Permalink to this headline">Â¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>##gff-version 3
##source-version NCBI C++ formatter 0.2
##date 2009-04-25
##Type DNA NC_008596.1
NC_008596.1	RefSeq	gene	12272	13301	.	+	.	locus_tag=MSMEG_0013;note=ferric%20enterobactin%20transport%20system%20permease%20protein%20FepG%3B%20this%20gene%20contains%20a%20frame%20shift%20which%20is%20not%20the%20result%20of%20sequencing%20error%3B%20identified%20by%20match%20to%20protein%20family%20HMM%20PF01032;pseudo=;db_xref=GeneID:4537201
NC_008596.1	RefSeq	gene	1137579	1138550	.	+	.	ID=NC_008596.1:speB;locus_tag=MSMEG_1072;db_xref=GeneID:4535378
NC_008596.1	RefSeq	CDS	1137579	1138547	.	+	0	ID=NC_008596.1:speB:unknown_transcript_1;Parent=NC_008596.1:speB;locus_tag=MSMEG_1072;EC_number=3.5.3.11;note=identified%20by%20match%20to%20protein%20family%20HMM%20PF00491%3B%20match%20to%20protein%20family%20HMM%20TIGR01230;transl_table=11;product=agmatinase;protein_id=YP_885468.1;db_xref=GI:118469242;db_xref=GeneID:4535378;exon_number=1
NC_008596.1	RefSeq	start_codon	1137579	1137581	.	+	0	ID=NC_008596.1:speB:unknown_transcript_1;Parent=NC_008596.1:speB;locus_tag=MSMEG_1072;EC_number=3.5.3.11;note=identified%20by%20match%20to%20protein%20family%20HMM%20PF00491%3B%20match%20to%20protein%20family%20HMM%20TIGR01230;transl_table=11;product=agmatinase;protein_id=YP_885468.1;db_xref=GI:118469242;db_xref=GeneID:4535378;exon_number=1
NC_008596.1	RefSeq	stop_codon	1138548	1138550	.	+	0	ID=NC_008596.1:speB:unknown_transcript_1;Parent=NC_008596.1:speB;locus_tag=MSMEG_1072;EC_number=3.5.3.11;note=identified%20by%20match%20to%20protein%20family%20HMM%20PF00491%3B%20match%20to%20protein%20family%20HMM%20TIGR01230;transl_table=11;product=agmatinase;protein_id=YP_885468.1;db_xref=GI:118469242;db_xref=GeneID:4535378;exon_number=1
NC_008596.1	RefSeq	gene	3597069	3598112	.	+	.	ID=NC_008596.1:speB;locus_tag=MSMEG_3535;db_xref=GeneID:4533678
NC_008596.1	RefSeq	CDS	3597069	3598109	.	+	0	ID=NC_008596.1:speB:unknown_transcript_2;Parent=NC_008596.1:speB;locus_tag=MSMEG_3535;EC_number=3.5.3.11;note=identified%20by%20match%20to%20protein%20family%20HMM%20PF00491%3B%20match%20to%20protein%20family%20HMM%20TIGR01230;transl_table=11;product=agmatinase;protein_id=YP_887838.1;db_xref=GI:118470943;db_xref=GeneID:4533678;exon_number=1
NC_008596.1	RefSeq	start_codon	3597069	3597071	.	+	0	ID=NC_008596.1:speB:unknown_transcript_2;Parent=NC_008596.1:speB;locus_tag=MSMEG_3535;EC_number=3.5.3.11;note=identified%20by%20match%20to%20protein%20family%20HMM%20PF00491%3B%20match%20to%20protein%20family%20HMM%20TIGR01230;transl_table=11;product=agmatinase;protein_id=YP_887838.1;db_xref=GI:118470943;db_xref=GeneID:4533678;exon_number=1
NC_008596.1	RefSeq	stop_codon	3598110	3598112	.	+	0	ID=NC_008596.1:speB:unknown_transcript_2;Parent=NC_008596.1:speB;locus_tag=MSMEG_3535;EC_number=3.5.3.11;note=identified%20by%20match%20to%20protein%20family%20HMM%20PF00491%3B%20match%20to%20protein%20family%20HMM%20TIGR01230;transl_table=11;product=agmatinase;protein_id=YP_887838.1;db_xref=GI:118470943;db_xref=GeneID:4533678;exon_number=1
NC_008596.1	RefSeq	gene	4460713	4461672	.	-	.	ID=NC_008596.1:speB;locus_tag=MSMEG_4374;db_xref=GeneID:4535424
NC_008596.1	RefSeq	CDS	4460716	4461672	.	-	0	ID=NC_008596.1:speB:unknown_transcript_3;Parent=NC_008596.1:speB;locus_tag=MSMEG_4374;EC_number=3.5.3.11;note=identified%20by%20match%20to%20protein%20family%20HMM%20PF00491%3B%20match%20to%20protein%20family%20HMM%20TIGR01230;transl_table=11;product=agmatinase;protein_id=YP_888649.1;db_xref=GI:118469662;db_xref=GeneID:4535424;exon_number=1
NC_008596.1	RefSeq	start_codon	4461670	4461672	.	-	0	ID=NC_008596.1:speB:unknown_transcript_3;Parent=NC_008596.1:speB;locus_tag=MSMEG_4374;EC_number=3.5.3.11;note=identified%20by%20match%20to%20protein%20family%20HMM%20PF00491%3B%20match%20to%20protein%20family%20HMM%20TIGR01230;transl_table=11;product=agmatinase;protein_id=YP_888649.1;db_xref=GI:118469662;db_xref=GeneID:4535424;exon_number=1
NC_008596.1	RefSeq	stop_codon	4460713	4460715	.	-	0	ID=NC_008596.1:speB:unknown_transcript_3;Parent=NC_008596.1:speB;locus_tag=MSMEG_4374;EC_number=3.5.3.11;note=identified%20by%20match%20to%20protein%20family%20HMM%20PF00491%3B%20match%20to%20protein%20family%20HMM%20TIGR01230;transl_table=11;product=agmatinase;protein_id=YP_888649.1;db_xref=GI:118469662;db_xref=GeneID:4535424;exon_number=1
NC_008596.1	RefSeq	gene	4539385	4540344	.	+	.	ID=NC_008596.1:speB;locus_tag=MSMEG_4459;db_xref=GeneID:4537057
NC_008596.1	RefSeq	CDS	4539385	4540341	.	+	0	ID=NC_008596.1:speB:unknown_transcript_4;Parent=NC_008596.1:speB;locus_tag=MSMEG_4459;EC_number=3.5.3.11;note=identified%20by%20match%20to%20protein%20family%20HMM%20PF00491%3B%20match%20to%20protein%20family%20HMM%20TIGR01230;transl_table=11;product=agmatinase;protein_id=YP_888732.1;db_xref=GI:118472833;db_xref=GeneID:4537057;exon_number=1
NC_008596.1	RefSeq	start_codon	4539385	4539387	.	+	0	ID=NC_008596.1:speB:unknown_transcript_4;Parent=NC_008596.1:speB;locus_tag=MSMEG_4459;EC_number=3.5.3.11;note=identified%20by%20match%20to%20protein%20family%20HMM%20PF00491%3B%20match%20to%20protein%20family%20HMM%20TIGR01230;transl_table=11;product=agmatinase;protein_id=YP_888732.1;db_xref=GI:118472833;db_xref=GeneID:4537057;exon_number=1
NC_008596.1	RefSeq	stop_codon	4540342	4540344	.	+	0	ID=NC_008596.1:speB:unknown_transcript_4;Parent=NC_008596.1:speB;locus_tag=MSMEG_4459;EC_number=3.5.3.11;note=identified%20by%20match%20to%20protein%20family%20HMM%20PF00491%3B%20match%20to%20protein%20family%20HMM%20TIGR01230;transl_table=11;product=agmatinase;protein_id=YP_888732.1;db_xref=GI:118472833;db_xref=GeneID:4537057;exon_number=1
</pre></div>
</div>
</div>
<div class="section" id="id31">
<h3>Import<a class="headerlink" href="#id31" title="Permalink to this headline">Â¶</a></h3>
<p>For &#8220;gene&#8221; featuretypes, we want to use the &#8220;db_xref&#8221; field as the primary key;
all other featuretypes will be incremented versions of the featuretype
(&#8220;CDS_1&#8221;, &#8220;CDS_2&#8221;, etc).</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fn</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span><span class="s">&#39;ncbi_gff3.txt&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">create_db</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="s">&quot;:memory:&quot;</span><span class="p">,</span> <span class="n">id_spec</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;gene&#39;</span><span class="p">:</span> <span class="s">&#39;db_xref&#39;</span><span class="p">})</span>
</pre></div>
</div>
</div>
<div class="section" id="id32">
<h3>Access<a class="headerlink" href="#id32" title="Permalink to this headline">Â¶</a></h3>
<p>Genes can now be accessed by the db_xref:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">db</span><span class="p">[</span><span class="s">&#39;GeneID:4537201&#39;</span><span class="p">]</span>
<span class="go">&lt;Feature gene (NC_008596.1:12272-13301[+]) at 0x...&gt;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="wormbase-gff2-alt-txt">
<span id="id33"></span><h2>wormbase_gff2_alt.txt<a class="headerlink" href="#wormbase-gff2-alt-txt" title="Permalink to this headline">Â¶</a></h2>
<p>This file has no gene_id or transcript_id fields; it appears to use &#8220;CDS&#8221; as
a gene-level kind of object.  So we can use a transform function to add
&#8220;gene_id&#8221; and &#8220;transcript_id&#8221; fields to all non-CDS features so that the file
conforms to a GTF standard and gene extents can be inferred.  Furthermore, by
default <a class="reference internal" href="api.html#module-gffutils" title="gffutils"><code class="xref py py-mod docutils literal"><span class="pre">gffutils</span></code></a> uses <code class="docutils literal"><span class="pre">'exon'</span></code> as the default feature type to merge into
genes.  Here, we need to specify <code class="docutils literal"><span class="pre">gtf_subfeature='coding_exon'</span></code>.</p>
<div class="section" id="id34">
<h3>File contents<a class="headerlink" href="#id34" title="Permalink to this headline">Â¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>Remanei_genome	Genomic_canonical	region	1	7816	.	+	.	Sequence &quot;Contig1020&quot;;
Contig102	WU_MERGED	CDS	1629	3377	.	-	.	CDS &quot;cr01.sctg102.wum.2.1&quot;
Contig102	WU_MERGED	coding_exon	2927	3377	.	-	.	CDS &quot;cr01.sctg102.wum.2.1&quot;
Contig102	WU_MERGED	coding_exon	2474	2875	.	-	.	CDS &quot;cr01.sctg102.wum.2.1&quot;
Contig102	WU_MERGED	coding_exon	1928	2430	.	-	.	CDS &quot;cr01.sctg102.wum.2.1&quot;
Contig102	WU_MERGED	coding_exon	1629	1883	.	-	.	CDS &quot;cr01.sctg102.wum.2.1&quot;
Contig102	WU_MERGED	intron	2876	2926	.	-	.	CDS &quot;cr01.sctg102.wum.2.1&quot;
Contig102	WU_MERGED	intron	2431	2473	.	-	.	CDS &quot;cr01.sctg102.wum.2.1&quot;
Contig102	WU_MERGED	intron	1884	1927	.	-	.	CDS &quot;cr01.sctg102.wum.2.1&quot;
</pre></div>
</div>
</div>
<div class="section" id="id35">
<h3>Import<a class="headerlink" href="#id35" title="Permalink to this headline">Â¶</a></h3>
<p>The transform function to manipulate the attribute dictionary: add &#8220;gene_id&#8221;
and &#8220;transcript_id&#8221; attributes to intron and coding_exon features:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">transform</span><span class="p">(</span><span class="n">f</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">f</span><span class="o">.</span><span class="n">featuretype</span> <span class="ow">in</span> <span class="p">[</span><span class="s">&#39;coding_exon&#39;</span><span class="p">,</span> <span class="s">&#39;intron&#39;</span><span class="p">]:</span>
<span class="gp">... </span>        <span class="n">_id</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;CDS&#39;</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
<span class="gp">... </span>        <span class="n">f</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;gene_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">_id</span><span class="p">]</span>
<span class="gp">... </span>        <span class="n">f</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;transcript_id&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">_id</span> <span class="o">+</span> <span class="s">&#39;_transcript&#39;</span><span class="p">]</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">f</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fn</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span><span class="s">&#39;wormbase_gff2_alt.txt&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">create_db</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span> <span class="s">&quot;:memory:&quot;</span><span class="p">,</span> <span class="n">id_spec</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;gene&#39;</span><span class="p">:</span> <span class="s">&#39;gene_id&#39;</span><span class="p">,</span> <span class="s">&#39;transcript&#39;</span><span class="p">:</span> <span class="s">&#39;transcript_id&#39;</span><span class="p">},</span> <span class="n">transform</span><span class="o">=</span><span class="n">transform</span><span class="p">,</span> <span class="n">gtf_subfeature</span><span class="o">=</span><span class="s">&#39;coding_exon&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="id36">
<h3>Access<a class="headerlink" href="#id36" title="Permalink to this headline">Â¶</a></h3>
<p>Note that the inferred genes have a source of &#8220;gffutils_derived&#8221;:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">db</span><span class="p">[</span><span class="s">&quot;cr01.sctg102.wum.2.1&quot;</span><span class="p">])</span>  
<span class="go">Contig102       gffutils_derived        gene    1629    3377    .       -       .       gene_id &quot;cr01.sctg102.wum.2.1&quot;;</span>
</pre></div>
</div>
<p>Get a report of the childrent of the gene:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">f</span> <span class="ow">in</span> <span class="n">db</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="s">&quot;cr01.sctg102.wum.2.1&quot;</span><span class="p">,</span> <span class="n">order_by</span><span class="o">=</span><span class="s">&#39;featuretype&#39;</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">print</span><span class="p">(</span><span class="s">&#39;{0.featuretype:&gt;12}: {0.id}&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">f</span><span class="p">))</span>
<span class="go"> coding_exon: coding_exon_1</span>
<span class="go"> coding_exon: coding_exon_2</span>
<span class="go"> coding_exon: coding_exon_3</span>
<span class="go"> coding_exon: coding_exon_4</span>
<span class="go">      intron: intron_1</span>
<span class="go">      intron: intron_2</span>
<span class="go">      intron: intron_3</span>
<span class="go">  transcript: cr01.sctg102.wum.2.1_transcript</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="wormbase-gff2-txt">
<span id="id37"></span><h2>wormbase_gff2.txt<a class="headerlink" href="#wormbase-gff2-txt" title="Permalink to this headline">Â¶</a></h2>
<p>This file is problematic because of inconsistent formatting: the <code class="docutils literal"><span class="pre">SAGE_tag</span></code>
features have a different attributes format than the rest of the features.  So
we need to do some work with the dialect used for parsing attributes into
dictionaries.</p>
<div class="section" id="id38">
<h3>File contents<a class="headerlink" href="#id38" title="Permalink to this headline">Â¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre>I	Genomic_canonical	region	1	2679	.	+	.	Sequence &quot;cTel33B&quot; ; Note &quot;Clone cTel33B; Genbank AC199162&quot; ; Note &quot;Clone cTel33B; Genbank AC199162&quot;
I	Coding_transcript	Transcript	12759582	12764949	.	-	.	Transcript &quot;B0019.1&quot; ; WormPep &quot;WP:CE40797&quot; ; Note &quot;amx-2&quot; ; Prediction_status &quot;Partially_confirmed&quot; ; Gene &quot;WBGene00000138&quot; ; CDS &quot;B0019.1&quot; ; WormPep &quot;WP:CE40797&quot; ; Note &quot;amx-2&quot; ; Prediction_status &quot;Partially_confirmed&quot; ; Gene &quot;WBGene00000138&quot;
I	Coding_transcript	intron	12759829	12759948	.	-	.	Transcript &quot;B0019.1&quot; ; Confirmed_EST EC034652
I	Coding_transcript	intron	12760014	12760226	.	-	.	Transcript &quot;B0019.1&quot; ; Confirmed_EST EC034652
I	Coding_transcript	intron	12760320	12760364	.	-	.	Transcript &quot;B0019.1&quot; ; Confirmed_EST yk1054h04.3
I	Coding_transcript	intron	12760495	12760833	.	-	.	Transcript &quot;B0019.1&quot; ; Confirmed_EST EC027594
I	Coding_transcript	intron	12760905	12761171	.	-	.	Transcript &quot;B0019.1&quot; ; Confirmed_EST EC027594
I	Coding_transcript	intron	12761517	12761798	.	-	.	Transcript &quot;B0019.1&quot; ; Confirmed_EST EC027594
I	Coding_transcript	intron	12761954	12762126	.	-	.	Transcript &quot;B0019.1&quot; ; Confirmed_EST yk262g9.5
I	Coding_transcript	intron	12762269	12762647	.	-	.	Transcript &quot;B0019.1&quot; ; Confirmed_EST yk262g9.5
I	Coding_transcript	intron	12762807	12763111	.	-	.	Transcript &quot;B0019.1&quot; ; Confirmed_EST yk1056c07.5
I	Coding_transcript	intron	12763250	12763447	.	-	.	Transcript &quot;B0019.1&quot; ; Confirmed_EST yk1056c07.5
I	Coding_transcript	intron	12763656	12763728	.	-	.	Transcript &quot;B0019.1&quot; ; Confirmed_EST yk1056c07.5
I	Coding_transcript	intron	12763883	12763978	.	-	.	Transcript &quot;B0019.1&quot; ; Confirmed_EST yk1054h04.5 ; Confirmed_EST OSTF088D9_1
I	Coding_transcript	intron	12764103	12764290	.	-	.	Transcript &quot;B0019.1&quot; ; Confirmed_EST yk1054h04.5 ; Confirmed_EST OSTF088D9_1
I	Coding_transcript	intron	12764472	12764811	.	-	.	Transcript &quot;B0019.1&quot; ; Confirmed_EST yk1054h04.5 ; Confirmed_EST OSTF088D9_1
I	Coding_transcript	exon	12759582	12759828	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	exon	12759949	12760013	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	exon	12760227	12760319	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	exon	12760365	12760494	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	exon	12760834	12760904	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	exon	12761172	12761516	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	exon	12761799	12761953	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	exon	12762127	12762268	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	exon	12762648	12762806	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	exon	12763112	12763249	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	exon	12763448	12763655	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	exon	12763729	12763882	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	exon	12763979	12764102	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	exon	12764291	12764471	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	exon	12764812	12764949	.	-	.	Transcript &quot;B0019.1&quot;
I	SAGE_tag_unambiguously_mapped	SAGE_tag	12761492	12761512	.	-	.	Sequence SAGE:aacggagccgtacacgc;count 5;Gene amx-2;Transcript B0019.1
I	SAGE_tag_most_three_prime	SAGE_tag	12761499	12761512	.	-	.	Sequence SAGE:aacggagccg;count 9;Gene amx-2;Transcript B0019.1
I	mass_spec_genome	translated_nucleotide_match	12761920	12761953	.	-	.	Target &quot;Mass_spec_peptide:MSP:FADFSPLDVSDVNFATDDLAK&quot; 10 21 ; Note &quot;MSP:FADFSPLDVSDVNFATDDLAK&quot; ; Protein_matches &quot;WP:CE40797&quot; ; CDS_matches &quot;B0019.1&quot; ; Times_observed &quot;3&quot;
I	mass_spec_genome	translated_nucleotide_match	12762127	12762155	.	-	.	Target &quot;Mass_spec_peptide:MSP:FADFSPLDVSDVNFATDDLAK&quot; 1 10 ; Note &quot;MSP:FADFSPLDVSDVNFATDDLAK&quot; ; Protein_matches &quot;WP:CE40797&quot; ; CDS_matches &quot;B0019.1&quot; ; Times_observed &quot;3&quot;
I	mass_spec_genome	translated_nucleotide_match	12763506	12763559	.	-	.	Target &quot;Mass_spec_peptide:MSP:FGHGQSLLAQGGMNEVVR&quot; 1 18 ; Note &quot;MSP:FGHGQSLLAQGGMNEVVR&quot; ; Protein_matches &quot;WP:CE40797&quot; ; CDS_matches &quot;B0019.1&quot; ; Times_observed &quot;1&quot;
I	SAGE_tag_unambiguously_mapped	SAGE_tag	12763533	12763553	.	-	.	Sequence SAGE:ggcagagtcttttggca;count 1;Gene amx-2;Transcript B0019.1
I	mass_spec_genome	translated_nucleotide_match	12764361	12764411	.	-	.	Target &quot;Mass_spec_peptide:MSP:NIQQNRPGLSVLVLEAR&quot; 1 17 ; Note &quot;MSP:NIQQNRPGLSVLVLEAR&quot; ; Protein_matches &quot;WP:CE40797&quot; ; CDS_matches &quot;B0019.1&quot; ; Times_observed &quot;2&quot;
I	GenePair_STS	PCR_product	12762449	12764118	.	+	.	PCR_product &quot;sjj_B0019.1&quot;
I	Expr_profile	experimental_result_region	12762449	12764118	.	+	.	Expr_profile &quot;B0019.1&quot;
I	Allele	SNP	12764272	12764272	.	+	.	Variation &quot;snp_B0019[1]&quot; ; Interpolated_map_position &quot;14.003&quot; ; ; RFLP &quot;No&quot;
I	Promoterome	PCR_product	12764938	12766937	.	+	.	PCR_product &quot;p_B0019.1_93&quot;
I	Oligo_set	reagent	12759745	12761589	.	-	.	Oligo_set &quot;Aff_B0019.1&quot;
I	Orfeome	PCR_product	12759747	12764936	.	-	.	PCR_product &quot;mv_B0019.1&quot; ; Amplified 1 ; Amplified 1
I	Coding_transcript	three_prime_UTR	12759582	12759744	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	coding_exon	12759745	12759828	.	-	0	Transcript &quot;B0019.1&quot; ; CDS &quot;B0019.1&quot;
I	Coding_transcript	coding_exon	12759949	12760013	.	-	2	Transcript &quot;B0019.1&quot; ; CDS &quot;B0019.1&quot;
I	Coding_transcript	coding_exon	12760227	12760319	.	-	2	Transcript &quot;B0019.1&quot; ; CDS &quot;B0019.1&quot;
I	Coding_transcript	coding_exon	12760365	12760494	.	-	0	Transcript &quot;B0019.1&quot; ; CDS &quot;B0019.1&quot;
I	Coding_transcript	coding_exon	12760834	12760904	.	-	2	Transcript &quot;B0019.1&quot; ; CDS &quot;B0019.1&quot;
I	Coding_transcript	coding_exon	12761172	12761516	.	-	2	Transcript &quot;B0019.1&quot; ; CDS &quot;B0019.1&quot;
I	Coding_transcript	coding_exon	12761799	12761953	.	-	1	Transcript &quot;B0019.1&quot; ; CDS &quot;B0019.1&quot;
I	Coding_transcript	coding_exon	12762127	12762268	.	-	2	Transcript &quot;B0019.1&quot; ; CDS &quot;B0019.1&quot;
I	Coding_transcript	coding_exon	12762648	12762806	.	-	2	Transcript &quot;B0019.1&quot; ; CDS &quot;B0019.1&quot;
I	Coding_transcript	coding_exon	12763112	12763249	.	-	2	Transcript &quot;B0019.1&quot; ; CDS &quot;B0019.1&quot;
I	Coding_transcript	coding_exon	12763448	12763655	.	-	0	Transcript &quot;B0019.1&quot; ; CDS &quot;B0019.1&quot;
I	Coding_transcript	coding_exon	12763729	12763882	.	-	1	Transcript &quot;B0019.1&quot; ; CDS &quot;B0019.1&quot;
I	Coding_transcript	coding_exon	12763979	12764102	.	-	2	Transcript &quot;B0019.1&quot; ; CDS &quot;B0019.1&quot;
I	Coding_transcript	coding_exon	12764291	12764471	.	-	0	Transcript &quot;B0019.1&quot; ; CDS &quot;B0019.1&quot;
I	Coding_transcript	five_prime_UTR	12764938	12764949	.	-	.	Transcript &quot;B0019.1&quot;
I	Coding_transcript	coding_exon	12764812	12764937	.	-	0	Transcript &quot;B0019.1&quot; ; CDS &quot;B0019.1&quot;
X	SAGE_tag	SAGE_tag	6819353	6819366	.	+	.	Sequence SAGE:aacggagccg;count 9;Gene amx-2;Transcript B0019.1
X 	gene 	processed_transcript 	944828 	948883 	.	-	.    	Gene &quot;WBGene00004893&quot;
</pre></div>
</div>
</div>
<div class="section" id="id39">
<h3>Import<a class="headerlink" href="#id39" title="Permalink to this headline">Â¶</a></h3>
<p>We will need <code class="docutils literal"><span class="pre">force_dialect_check=True</span></code> in this case if we want to be able to have
the parents of the <code class="docutils literal"><span class="pre">SAGE_tag</span></code> features be recognized, because the attributes
for <code class="docutils literal"><span class="pre">SAGE_tag</span></code> features are formatted differently.  That is, they don&#8217;t have
quoted values, and the separating semicolon does not have a space on either
side as is the case for the other features.</p>
<p>The default of <code class="docutils literal"><span class="pre">False</span></code> means that only the first few lines are checked to
determine the dialect to be used for all lines.  But that won&#8217;t work here
because of the inconsistent formatting.  So we have to take the more
time-consuming approach of checking every dialect in order to figure out how to
parse them into a dictionary.</p>
<p>Transform function to create appropriate &#8220;Parent&#8221; attributes:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">def</span> <span class="nf">transform</span><span class="p">(</span><span class="n">f</span><span class="p">):</span>
<span class="gp">... </span>    <span class="k">if</span> <span class="n">f</span><span class="o">.</span><span class="n">featuretype</span> <span class="o">==</span> <span class="s">&#39;Transcript&#39;</span><span class="p">:</span>
<span class="gp">... </span>        <span class="n">f</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;Parent&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;Gene&#39;</span><span class="p">]</span>
<span class="gp">... </span>    <span class="k">else</span><span class="p">:</span>
<span class="gp">... </span>        <span class="k">try</span><span class="p">:</span>
<span class="gp">... </span>            <span class="n">f</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;Parent&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;Transcript&#39;</span><span class="p">]</span>
<span class="gp">... </span>        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
<span class="gp">... </span>            <span class="k">pass</span>
<span class="gp">... </span>    <span class="k">return</span> <span class="n">f</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fn</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span><span class="s">&#39;wormbase_gff2.txt&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">create_db</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span>
<span class="gp">... </span><span class="s">&quot;:memory:&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">transform</span><span class="o">=</span><span class="n">transform</span><span class="p">,</span>
<span class="gp">... </span><span class="n">id_spec</span><span class="o">=</span><span class="p">{</span><span class="s">&#39;Transcript&#39;</span><span class="p">:</span> <span class="s">&quot;Transcript&quot;</span><span class="p">},</span>
<span class="gp">... </span><span class="n">force_gff</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
<span class="gp">... </span><span class="n">force_dialect_check</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
<span class="gp">... </span><span class="n">keep_order</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<p>The dialect for the database will be None:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">assert</span> <span class="n">db</span><span class="o">.</span><span class="n">dialect</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">,</span> <span class="n">db</span><span class="o">.</span><span class="n">dialect</span>
</pre></div>
</div>
<p>For cases like this, we should probably construct our own dialect to force all
attributes to have the same format.  To help with this, we can use the
<code class="xref py py-func docutils literal"><span class="pre">helpers.infer_dialect()</span></code> function by providing attributes:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">gffutils</span> <span class="kn">import</span> <span class="n">helpers</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dialect</span> <span class="o">=</span> <span class="n">helpers</span><span class="o">.</span><span class="n">infer_dialect</span><span class="p">([</span>
<span class="gp">... </span><span class="s">&#39;Transcript &quot;B0019.1&quot; ; WormPep &quot;WP:CE40797&quot; ; Note &quot;amx-2&quot; ; Prediction_status &quot;Partially_confirmed&quot; ; Gene &quot;WBGene00000138&quot; ; CDS &quot;B0019.1&quot; ; WormPep &quot;WP:CE40797&quot; ; Note &quot;amx-2&quot; ; Prediction_status &quot;Partially_confirmed&quot; ; Gene &quot;WBGene00000138&quot;&#39;</span><span class="p">,</span>
<span class="gp">... </span><span class="p">])</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">db</span><span class="o">.</span><span class="n">dialect</span> <span class="o">=</span> <span class="n">dialect</span>
</pre></div>
</div>
</div>
<div class="section" id="id40">
<h3>Access<a class="headerlink" href="#id40" title="Permalink to this headline">Â¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">t</span> <span class="o">=</span> <span class="n">db</span><span class="p">[</span><span class="s">&quot;B0019.1&quot;</span><span class="p">]</span>
</pre></div>
</div>
<p>Since we&#8217;ve set the dialect for the database, any features returned from the
database should follow that dialect:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">t</span><span class="p">)</span>  
<span class="go">I       Coding_transcript       Transcript      12759582        12764949        .       -       .       Transcript &quot;B0019.1&quot; ; WormPep &quot;WP:CE40797&quot; ; WormPep &quot;WP:CE40797&quot; ; Note &quot;amx-2&quot; ; Note &quot;amx-2&quot; ; Prediction_status &quot;Partially_confirmed&quot; ; Prediction_status &quot;Partially_confirmed&quot; ; Gene &quot;WBGene00000138&quot; ; Gene &quot;WBGene00000138&quot; ; CDS &quot;B0019.1&quot; ; Parent &quot;WBGene00000138&quot; ; Parent &quot;WBGene00000138&quot;</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">db</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">featuretype</span><span class="o">=</span><span class="s">&#39;exon&#39;</span><span class="p">)))</span>
<span class="go">15</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">db</span><span class="p">[</span><span class="s">&#39;SAGE_tag_1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">attributes</span><span class="p">[</span><span class="s">&#39;Transcript&#39;</span><span class="p">]</span>
<span class="go">[&#39;B0019.1&#39;]</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="nb">len</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">db</span><span class="o">.</span><span class="n">children</span><span class="p">(</span><span class="n">t</span><span class="p">,</span> <span class="n">featuretype</span><span class="o">=</span><span class="s">&#39;SAGE_tag&#39;</span><span class="p">)))</span>
<span class="go">4</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="gencode-v19-gtf">
<span id="gencode-v190-gtf"></span><h2>gencode-v19.gtf<a class="headerlink" href="#gencode-v19-gtf" title="Permalink to this headline">Â¶</a></h2>
<p>This example file contains the first gene from the Gencode v19 annotations.
It&#8217;s in GTF format, but in contrast to an on-spec GTF file, which only has
exon, CDS, UTRs, and start/stop codon features, this file already has genes and
transcripts on separate lines.  This means we can avoid the gene and transcript
inference, which saves time.</p>
<div class="highlight-python"><div class="highlight"><pre>##description: evidence-based annotation of the human genome (GRCh37), version 19 (Ensembl 74)
##provider: GENCODE
##contact: gencode@sanger.ac.uk
##format: gtf
##date: 2013-12-05
chr1	HAVANA	gene	11869	14412	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENSG00000223972.4&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1&quot;; level 2; havana_gene &quot;OTTHUMG00000000961.2&quot;;
chr1	HAVANA	transcript	11869	14409	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000456328.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;processed_transcript&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-002&quot;; level 2; tag &quot;basic&quot;; havana_gene &quot;OTTHUMG00000000961.2&quot;; havana_transcript &quot;OTTHUMT00000362751.1&quot;;
chr1	HAVANA	exon	11869	12227	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000456328.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;processed_transcript&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-002&quot;; exon_number 1;  exon_id &quot;ENSE00002234944.1&quot;;  level 2; tag &quot;basic&quot;; havana_gene &quot;OTTHUMG00000000961.2&quot;; havana_transcript &quot;OTTHUMT00000362751.1&quot;;
chr1	HAVANA	exon	12613	12721	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000456328.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;processed_transcript&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-002&quot;; exon_number 2;  exon_id &quot;ENSE00003582793.1&quot;;  level 2; tag &quot;basic&quot;; havana_gene &quot;OTTHUMG00000000961.2&quot;; havana_transcript &quot;OTTHUMT00000362751.1&quot;;
chr1	HAVANA	exon	13221	14409	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000456328.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;processed_transcript&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-002&quot;; exon_number 3;  exon_id &quot;ENSE00002312635.1&quot;;  level 2; tag &quot;basic&quot;; havana_gene &quot;OTTHUMG00000000961.2&quot;; havana_transcript &quot;OTTHUMT00000362751.1&quot;;
chr1	ENSEMBL	transcript	11872	14412	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000515242.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-201&quot;; level 3; havana_gene &quot;OTTHUMG00000000961.2&quot;;
chr1	ENSEMBL	exon	11872	12227	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000515242.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-201&quot;; exon_number 1;  exon_id &quot;ENSE00002234632.1&quot;;  level 3; havana_gene &quot;OTTHUMG00000000961.2&quot;;
chr1	ENSEMBL	exon	12613	12721	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000515242.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-201&quot;; exon_number 2;  exon_id &quot;ENSE00003608237.1&quot;;  level 3; havana_gene &quot;OTTHUMG00000000961.2&quot;;
chr1	ENSEMBL	exon	13225	14412	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000515242.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-201&quot;; exon_number 3;  exon_id &quot;ENSE00002306041.1&quot;;  level 3; havana_gene &quot;OTTHUMG00000000961.2&quot;;
chr1	ENSEMBL	transcript	11874	14409	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000518655.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-202&quot;; level 3; havana_gene &quot;OTTHUMG00000000961.2&quot;;
chr1	ENSEMBL	exon	11874	12227	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000518655.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-202&quot;; exon_number 1;  exon_id &quot;ENSE00002269724.1&quot;;  level 3; havana_gene &quot;OTTHUMG00000000961.2&quot;;
chr1	ENSEMBL	exon	12595	12721	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000518655.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-202&quot;; exon_number 2;  exon_id &quot;ENSE00002270865.1&quot;;  level 3; havana_gene &quot;OTTHUMG00000000961.2&quot;;
chr1	ENSEMBL	exon	13403	13655	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000518655.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-202&quot;; exon_number 3;  exon_id &quot;ENSE00002216795.1&quot;;  level 3; havana_gene &quot;OTTHUMG00000000961.2&quot;;
chr1	ENSEMBL	exon	13661	14409	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000518655.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-202&quot;; exon_number 4;  exon_id &quot;ENSE00002303382.1&quot;;  level 3; havana_gene &quot;OTTHUMG00000000961.2&quot;;
chr1	HAVANA	transcript	12010	13670	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000450305.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-001&quot;; level 2; ont &quot;PGO:0000005&quot;; ont &quot;PGO:0000019&quot;; havana_gene &quot;OTTHUMG00000000961.2&quot;; havana_transcript &quot;OTTHUMT00000002844.2&quot;;
chr1	HAVANA	exon	12010	12057	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000450305.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-001&quot;; exon_number 1;  exon_id &quot;ENSE00001948541.1&quot;;  level 2; ont &quot;PGO:0000005&quot;; ont &quot;PGO:0000019&quot;; havana_gene &quot;OTTHUMG00000000961.2&quot;; havana_transcript &quot;OTTHUMT00000002844.2&quot;;
chr1	HAVANA	exon	12179	12227	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000450305.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-001&quot;; exon_number 2;  exon_id &quot;ENSE00001671638.2&quot;;  level 2; ont &quot;PGO:0000005&quot;; ont &quot;PGO:0000019&quot;; havana_gene &quot;OTTHUMG00000000961.2&quot;; havana_transcript &quot;OTTHUMT00000002844.2&quot;;
chr1	HAVANA	exon	12613	12697	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000450305.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-001&quot;; exon_number 3;  exon_id &quot;ENSE00001758273.2&quot;;  level 2; ont &quot;PGO:0000005&quot;; ont &quot;PGO:0000019&quot;; havana_gene &quot;OTTHUMG00000000961.2&quot;; havana_transcript &quot;OTTHUMT00000002844.2&quot;;
chr1	HAVANA	exon	12975	13052	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000450305.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-001&quot;; exon_number 4;  exon_id &quot;ENSE00001799933.2&quot;;  level 2; ont &quot;PGO:0000005&quot;; ont &quot;PGO:0000019&quot;; havana_gene &quot;OTTHUMG00000000961.2&quot;; havana_transcript &quot;OTTHUMT00000002844.2&quot;;
chr1	HAVANA	exon	13221	13374	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000450305.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-001&quot;; exon_number 5;  exon_id &quot;ENSE00001746346.2&quot;;  level 2; ont &quot;PGO:0000005&quot;; ont &quot;PGO:0000019&quot;; havana_gene &quot;OTTHUMG00000000961.2&quot;; havana_transcript &quot;OTTHUMT00000002844.2&quot;;
chr1	HAVANA	exon	13453	13670	.	+	.	gene_id &quot;ENSG00000223972.4&quot;; transcript_id &quot;ENST00000450305.2&quot;; gene_type &quot;pseudogene&quot;; gene_status &quot;KNOWN&quot;; gene_name &quot;DDX11L1&quot;; transcript_type &quot;transcribed_unprocessed_pseudogene&quot;; transcript_status &quot;KNOWN&quot;; transcript_name &quot;DDX11L1-001&quot;; exon_number 6;  exon_id &quot;ENSE00001863096.1&quot;;  level 2; ont &quot;PGO:0000005&quot;; ont &quot;PGO:0000019&quot;; havana_gene &quot;OTTHUMG00000000961.2&quot;; havana_transcript &quot;OTTHUMT00000002844.2&quot;;
</pre></div>
</div>
<div class="section" id="id41">
<h3>Import<a class="headerlink" href="#id41" title="Permalink to this headline">Â¶</a></h3>
<p>We can use <code class="docutils literal"><span class="pre">disable_infer_genes=True</span></code> and <code class="docutils literal"><span class="pre">disable_infer_transcripts=True</span></code> to
disable the gene and transcript inference.  This will dramatically speed up
import, but will result in identical results:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">fn</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">example_filename</span><span class="p">(</span><span class="s">&#39;gencode-v19.gtf&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">db</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">create_db</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span>
<span class="gp">... </span><span class="s">&quot;:memory:&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">keep_order</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
<span class="gp">... </span><span class="n">disable_infer_genes</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">disable_infer_transcripts</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="id42">
<h3>Access<a class="headerlink" href="#id42" title="Permalink to this headline">Â¶</a></h3>
<p>Access is the same as other files:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">db</span><span class="p">[</span><span class="s">&#39;ENSG00000223972.4&#39;</span><span class="p">]</span>
<span class="go">&lt;Feature gene (chr1:11869-14412[+]) at 0x...&gt;</span>
</pre></div>
</div>
<p>Ensure that inferring gene extent results in the same database &#8211; it just takes
longer to import:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">db2</span> <span class="o">=</span> <span class="n">gffutils</span><span class="o">.</span><span class="n">create_db</span><span class="p">(</span><span class="n">fn</span><span class="p">,</span>
<span class="gp">... </span><span class="s">&quot;:memory:&quot;</span><span class="p">,</span>
<span class="gp">... </span><span class="n">keep_order</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span>
<span class="gp">... </span><span class="n">disable_infer_genes</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">disable_infer_transcripts</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">db</span><span class="o">.</span><span class="n">all_features</span><span class="p">(),</span> <span class="n">db2</span><span class="o">.</span><span class="n">all_features</span><span class="p">()):</span>
<span class="gp">... </span>    <span class="k">assert</span> <span class="n">i</span> <span class="o">==</span> <span class="n">j</span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="api.html" class="btn btn-neutral float-right" title="API" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="attributes.html" class="btn btn-neutral" title="Attributes" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2013, Ryan Dale.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.8.7.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>